---
label: Quality Trimming
icon: codescan-checkmark
order: 6
---

# Quality Trimming Sequence Data
===  :icon-checklist: You will need
- at least 2 cores/threads available
- paired-end b/gzipped fastq sequence files
===

Raw sequences are not suitable for downstream analyses. They have sequencing adapters,
index sequences, regions of poor quality, etc. The first step of any genetic sequence
analyses is to remove these adapters and trim poor quality data. You can remove adapters
and quality trim sequences using the `trim` module:
```bash
harpy trim OPTIONS... 
```

## Running Options
| argument         | short name | type        | default | required | description                                                            |
|:-----------------|:----------:|:------------|:-------:|:--------:|:-----------------------------------------------------------------------|
| `--dir`          |    `-d`    | folder path |         | **yes**  | Directory with sequence alignments                                     |
| `--max-length`   |    `-l`    | integer     |   150   |    no    | Maximum length to trim sequences down to                               |
| `--extra-params` |    `-x`    | string      |         |    no    | Additional fastp parameters, in quotes                                 |
| `--threads`      |    `-t`    | integer     |    4    |    no    | Number of threads to use                                               |
| `--snakemake`    |    `-s`    | string      |         |    no    | Additional Snakemake options, in quotes ([more info](../getstarted.md/#adding-additional-snakamake-parameters)) |
| `--help`         |            |             |         |          | Show the module docstring                                              |

## FASTQ file format
There are a handful of "accepted" naming schemes for fastq file extensions, but Harpy only accepts a limited number of them, shown below.
The fastq files **must** be bgzipped or gzipped and be **consistent** with regards to the extensions and read-pair naming styles.
That is, you must only use `.fastq.gz` or only use `.fq.gz` for all files, and the same for `.1.`/`.2.` or `.F.`/`.R.` (adhere to a single row in the table below).
Notice that the read pair part differs from the [accepted fastq formats](readmapping.md/#fastq-file-format) for aligning reads.
=== acceptable formats

| forward-reverse notation | extension  | example forward          | example reverse         |
|:-------------------------|:-----------|:-------------------------|:------------------------|
| `.F` / `.R`                | `fastq.gz` | ` samplename.F.fastq.gz` | `samplename.R.fastq.gz` |
| `.F` / `.R`                | `fq.gz`    | `samplename.F.fq.gz`     | `samplename.R.fq.gz`    |
| `.1` / `.2`                | `fastq.gz` | `samplename.2.fastq.gz`  | `samplename.2.fastq.gz` |
| `.1` / `.2`                | `fq.gz`    | `samplename.1.fq.gz`     | `samplename.2.fq.gz`    |

===

---
## Trimming Workflow
[Fastp](https://github.com/OpenGene/fastp) is an ultra-fast all-in-one adapter remover, deduplicator, 
and quality trimmer. Harpy uses it to remove adapters, low-quality bases, and trim sequences down to a particular
length (default 150bp). Harpy uses the fastp overlap analysis to identify adapters for removal and a sliding window
approach (`--cut-right`) to identify low quality bases. The workflow is quite simple.

```mermaid
graph LR
    A([fastp trim]) --> B([create reports])
```

### harpy trim outputs
The `harpy trim` module creates a `Trimming` directory with the folder structure below. `Sample1` and `Sample2` are generic sample names for demonstration purposes.
```
Trimming/
├── Sample1.R1.fq.gz
├── Sample1.R2.fq.gz
├── Sample2.R1.fq.gz
├── Sample2.R2.fq.gz
└── logs
    ├── err
    │   ├── Sample1.log
    │   └── Sample2.log
    ├── html
    │   ├── Sample1.html
    │   └── Sample2.html
    ├── json
    │   ├── Sample1.fastp.json
    │   └── Sample2.fastp.json
    └── trim.report.html
```
| item                    | description                                                              |
|:------------------------|:-------------------------------------------------------------------------|
| `*.R1.fq.gz`            | quality trimmed forward reads of the samples                             |
| `*.R1.fq.gz`            | quality trimmed reverse reads of the samples                             |
| `logs/`                 | all debug/diagnostic files that aren't the trimmed reads `fastp` creates |
| `logs/err`              | what fastp prints to `stderr` when running                               |
| `logs/html`             | interactive html reports `fastp` creates from quality trimming           |
| `logs/json`             | json representation of the data `fastp` uses to create the html reports  |
| `logs/trim.report.html` | a report generated by `multiqc` summarizing the quality trimming results |