---
label: Trim
description: Quality trim haplotagged sequences with Harpy
icon: codescan-checkmark
order: 6
---

# :icon-codescan-checkmark: Quality Trim Sequences
===  :icon-checklist: You will need
- at least 2 cores/threads available
- paired-end b/gzipped fastq sequence files with the [proper naming convention](../dataformat/#naming-conventions)
===

Raw sequences are not suitable for downstream analyses. They have sequencing adapters,
index sequences, regions of poor quality, etc. The first step of any genetic sequence
analyses is to remove these adapters and trim poor quality data. You can remove adapters
and quality trim sequences using the `trim` module:

```bash usage
harpy trim OPTIONS... 
```

```bash example
harpy trim --dir Sequences_Raw/ --threads 20 
```

## :icon-terminal: Running Options
| argument         | short name | type        | default | required | description                                                                                     |
|:-----------------|:----------:|:------------|:-------:|:--------:|:------------------------------------------------------------------------------------------------|
| `--dir`          |    `-d`    | folder path |         | **yes**  | Directory with sequence alignments                                                              |
| `--max-length`   |    `-l`    | integer     |   150   |    no    | Maximum length to trim sequences down to                                                        |
| `--extra-params` |    `-x`    | string      |         |    no    | Additional fastp arguments, in quotes                                                           |
| `--threads`      |    `-t`    | integer     |    4    |    no    | Number of threads to use                                                                        |
| `--snakemake`    |    `-s`    | string      |         |    no    | Additional [Snakemake](../snakemake/#adding-snakamake-parameters) options, in quotes |
| `--quiet`        |    `-q`    | toggle      |         |    no    | Supressing Snakemake printing to console                                                        |
| `--help`         |            |             |         |          | Show the module docstring                                                                       |

---
## :icon-git-pull-request: Trimming Workflow
+++ :icon-git-merge: details
[Fastp](https://github.com/OpenGene/fastp) is an ultra-fast all-in-one adapter remover, deduplicator, 
and quality trimmer. Harpy uses it to remove adapters, low-quality bases, and trim sequences down to a particular
length (default 150bp). Harpy uses the fastp overlap analysis to identify adapters for removal and a sliding window
approach (`--cut-right`) to identify low quality bases. The workflow is quite simple.

```mermaid
graph LR
    A([fastp trim]) --> B([count barcodes])
    A --> C([create reports])
    B --> C
```

+++ :icon-file-directory: trimming output
The `harpy trim` module creates a `Trimming` directory with the folder structure below. `Sample1` and `Sample2` are generic sample names for demonstration purposes.
```
Trim/
├── Sample1.R1.fq.gz
├── Sample1.R2.fq.gz
├── Sample2.R1.fq.gz
├── Sample2.R2.fq.gz
├── reports
│   ├── Sample1.html
│   ├── Sample2.html
│   ├── summary.bx.valid.html
│   └── trim.report.html
└── logs
    ├── err
    │   ├── Sample1.log
    │   └── Sample2.log
    └── json
        ├── Sample1.fastp.json
        └── Sample2.fastp.json
```
| item                            | description                                                                        |
|:--------------------------------|:-----------------------------------------------------------------------------------|
| `*.R1.fq.gz`                    | quality trimmed forward reads of the samples                                       |
| `*.R1.fq.gz`                    | quality trimmed reverse reads of the samples                                       |
| `logs/`                         | all debug/diagnostic files that aren't the trimmed reads `fastp` creates           |
| `logs/err`                      | what fastp prints to `stderr` when running                                         |
| `reports/*.html`                | interactive html reports `fastp` creates from quality trimming                     |
| `reports/trim.report.html`      | a report generated by `multiqc` summarizing the quality trimming results           |
| `reports/summary.bx.valid.html` | a report detailing valid vs invalid barcodes and the segments causing invalidation |
| `logs/json`                     | json representation of the data `fastp` uses to create the html reports            |

+++ :icon-code-square: fastp arguments
The list of all `fastp` command line options is quite extensive and would
be cumbersome to print here. See the list of options in the [fastp documentation](https://github.com/OpenGene/fastp).

+++
