---
title: "Sequence Alignment Depth and Coverage"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    toc: no
---

```{r echo = FALSE, message = FALSE, warning = FALSE}
library(ggplot2)
library(dplyr)
library(tidyr)
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
covfile <- snakemake@input[["gencov"]]
# covfile <- "~/ShadHap1_C01.gencov"
gencov <- read.table(covfile, header = F)
if (ncol(gencov) == 5) {
  mult <- FALSE
  gencov <- gencov[, c(1,2,3,5)]
  colnames(gencov) <- c("contig", "start", "end", "alignments")
} else {
  mult <- TRUE
  gencov <- gencov[, c(1,2,3,6,7)]
  colnames(gencov) <- c("contig", "start", "end", "all", "barcodeOnly")
  gencov$barcodeOnly <- gencov$barcodeOnly * -1
}

# Find the 30 largest contigs
contigs <- group_by(gencov, contig) %>%
  summarize(size = max(end)) %>%
  arrange(desc(size))

# limit the data to only the 30 largest contigs
if (nrow(contigs) > 30){
    pltheight <- 50
    gencov <- filter(gencov, contig %in% contigs$contig[1:30])
} else {
    pltheight <- round(3 * (nrow(contigs)), digits = 0)
}

# pivot longer for generalized plotting
gencov <- pivot_longer(gencov, -(1:3), names_to = "alignment", values_to = "depth")

#gencov$midpoint <- (gencov$start + gencov$end) / 2

#fa.sizes <- read.table(snakemake@input[["faidx"]], header = F)[, 1:2]
#fa.size <- fa.sizes[order(fa.sizes$V2), ]
#fa.sizes <- read.table("~/genome.fasta.fai", header = F)[,1:2]
#colnames(fa.sizes) <- c("contig", "size")

sname <- gsub("\\.\\w?\\w?\\w?.?cov.gz", "", basename(covfile))

```
## Alignment Depth
#### Sample: `r sname`

Plotted below are (up to) the 30 largest contigs to show sequence alignment depth across
the genome. Each bar represents the alignment depth at a 10kb genomic interval, that is, the
number of reads that had a proper alignment in the 10kb interval. "Proper" refers to a read
not marked as a duplicate or flagged with the SAM `UNMAP`, `SECONDARY`,  or `QCFAIL` flags.
These values are derived by using `samtools bedcov -n`.

```{r echo = FALSE, message = FALSE, warning = FALSE, fig.height=pltheight, out.width = "100%"}
if (mult) {
  ggplot(gencov, aes(xmin = start, xmax = end, ymin = 0, ymax = depth, color = alignment, fill = alignment)) +
    geom_rect() +
    theme_minimal() +
    facet_wrap(~contig, ncol = 2, scales = "free_x") +
    scale_color_manual(values = c("#3f3f3f", "dodgerblue")) +
    scale_fill_manual(values = c("#3f3f3f", "dodgerblue")) +
    scale_y_continuous(labels = abs) +
    labs(x = "position (bp)", y = "depth", title = "Alignment Depth") +
    theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(), legend.position = "top") +
    guides(x = "none") -> p
} else {
  ggplot(gencov, aes(xmin = start, xmax = end, ymin = 0, ymax = depth)) +
    geom_rect(color = "#3f3f3f", fill = "#3f3f3f") +
    theme_minimal() +
    facet_wrap(~contig, ncol = 3, scales = "free_x") +
    labs(x = "position (bp)", y = "depth") +
    theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(), legend.position = "top") +
    guides(x = "none") -> p
}
print(p)
```