

```{r message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(magrittr)
```

```{r echo = FALSE}
fa.sizes <- read.table("~/genome.fasta.fai", header = F)[,1:2]
colnames(fa.sizes) <- c("contig", "size")
```

```{r echo = FALSE}
sv <- read.table("~/1.sv.stats", header = T)
sv2 <- read.table("~/2.sv.stats", skip = 1, header = F)
colnames(sv2) <- colnames(sv)
sv <- rbind(sv, sv2)
sv <- merge(x=sv,y=fa.sizes, by="contig")
```



```{r echo = FALSE, fig.height = 1.5}
for(i in unique(fa.sizes$contig)){
  sv.filt <- sv %>% filter(contig == i)
  sv_stats <- group_by(sv.filt, type) %>%  summarise(n = length(type))
  if(nrow(sv_stats) == 0){
    next
  }
  plt <- sv.filt %>% 
    ggplot(aes(y = as.factor(population), yend = as.factor(population), x = position_start, xend = position_end, color = as.factor(type))) +
    scale_color_manual(values=c("DEL"="black", "DUP"="violet", "INV"="dodgerblue","BND"="orange")) +
    geom_segment(alpha = 0.5, linewidth = 10) +  
    theme(
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank()
    ) +
    xlim(1, sv$size[1]) +
    xlab("Position (bp)") +
    ylab("Population") +
    labs(color="SV type") +
    ggtitle(i)
  print(plt)
}
```
