---
title: "Leviathan Structural Variants Summary"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    df_print: kable
    toc: no
---

Below is a series of tables and plots detailing the structural variants
identified by [LEVIATHAN](https://github.com/morispi/LEVIATHAN) [(preprint)](https://doi.org/10.1101/2021.03.25.437002). The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
```
```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(magrittr)
library(tidyr)
library(DT)
library(plotly)
library(circlize)
library(viridisLite)
library(tools)
library(ComplexHeatmap)
```
```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r echo = FALSE, results='asis'}
#faidx <- "/mnt/Win11/Users/pdime/Documents/Omega/Cornell/Haplotagging/genome.fasta.fai"
faidx <- snakemake@input[["faidx"]]
fa.sizes <- read.table(faidx, header = F)[,1:2] %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")
# limit the data to only the 30 largest contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]


#statsfiles <- c("/mnt/Win11/Users/pdime/Documents/Omega/Cornell/Haplotagging/1.sv.stats", "/mnt/Win11/Users/pdime/Documents/Omega/Cornell/Haplotagging/2.sv.stats")
statsfiles <- snakemake@input[["statsfiles"]]
sv <- do.call(rbind, lapply(statsfiles, function(x) read.table(x, header = T)))
#.sv <- sv
#.sv$population <- ifelse(.sv$population == 1, 3, 4)
#.sv2 <- sv
#.sv2$population <- ifelse(.sv2$population == 3, 5, 6)
#sv <- rbind(sv, .sv, .sv2)

if (nrow(sv) == 0) {
  cat("There are no variants detected in files:\n")
  cat(paste0("- `", statsfiles, "`"), sep = "\n")
  knitr::knit_exit()
}
```

## Various information {.tabset}
### summary info
This table shows the overall counts, average per contig, and standard deviation per contig for detected structural variants across populations.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
genstats <- sv %>%
  group_by(population, contig, type) %>%
  summarise(tot = n()) %>%
  group_by(population, type) %>%
  summarise(count = sum(tot), avg = round(mean(tot, na.rm = T),2 ), sd = round(sd(tot, na.rm = T), 2)) %>% 
  pivot_wider(names_from = type, values_from = c(count, avg, sd), names_glue = "{type}.{.value}", names_vary = "slowest")

nicecols <- toTitleCase(gsub("\\."," ", names(genstats)))

DT::datatable(
  genstats,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = nicecols
)
```

### all variants
This table details all the variants LEVIATHAN detected and the quantity/quality of evidence for those detections.
```{r message = FALSE, echo = FALSE}
DT::datatable(sv, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

```{r echo = FALSE, warnings= FALSE, message = FALSE}
sv$length <- as.numeric(gsub("\\.", "1", sv$length))
```

## Per-contig information {.tabset}
### counts per contig
This table details counts each type of structural variant for every contig for every population.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grpcounts <- sv %>%
  group_by(population, contig, type) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = population, values_from = count)

DT::datatable(grpcounts, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

### bp per contig
This table details the lengths (in bp) of SV's for every contig for every population. Breakends are not counted because they do not have a size.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grplens <- sv %>% 
  filter(type != "BND") %>%
  group_by(population, contig, type) %>%
  summarise(bp = sum(length)) %>%
  pivot_wider(names_from = population, values_from = bp)

DT::datatable(grplens, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

---

## SV size and location {.tabset}
The plots below are a visual representation of each SV where it was detected.
Contigs that are absent did not have any SV's detected on them. Within each 
contig, each population is plotted as a row to help visualize shared
variants between populations. Plotting only the first 30 largest contigs.

### Summary

```{r echo = FALSE, message = FALSE, warning = FALSE,fig.height = 1, fig.asp=0.35, fig.align='center'}
#populations <- c("red", "blue", "green","orange","violet")  # relevant for dev
populations <- unique(sv$population)
col.ramp <- viridis(length(populations))

plot(NULL ,xaxt='n',yaxt='n',bty='n',ylab='',xlab='', xlim=0:1, ylim=0:1)
legend("topleft", legend = populations, pch=15, pt.cex=3, cex=1, bty='n',
    col = col.ramp, ncol=5)
mtext("Populations", at=0.1, cex=1.5)
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
circosplot <- function(data, contigdata, svtype, trackheight, poplist, colormap){
  circos.clear()
  col_text <- "grey40"
  circos.par("track.height" = 0.8, gap.degree = 2, cell.padding = c(0, 0, 0, 0))
  circos.initialize(
    factors = contigdata$contig,
    xlim = matrix(c(rep(0, nrow(contigdata)), contigdata$size), ncol = 2)
  )

  # contigs
  circos.track(
    ylim = c(0, 1),
    bg.col = "grey90",
    bg.border = F,
    track.height = 0.06,
    panel.fun = function(x, y) {
      chr = CELL_META$sector.index
      xlim = CELL_META$xlim
      ylim = CELL_META$ylim
      circos.text(
        mean(xlim),
        mean(ylim),
        chr,
        cex = 0.6,
        col = col_text,
        facing="bending.inside",
        niceFacing = T
      )
    }
  )

  # x axis
  brk <- seq(0, 10, 0.5) * 10^7
  circos.track(
    ylim = c(0,1),
    track.index = get.current.track.index(),
    bg.border = F,
    panel.fun = function(x, y) {
      circos.axis(
        h = "top",
        major.at = brk,
        labels = round(brk/10^6, 1),
        labels.cex = 0.6,
        col = col_text,
        labels.col = col_text,
        lwd = 0.7,
        labels.facing = "clockwise"
      )
    }
  )

  # plot the track
  for(i in 1:length(poplist)){
    svquery = data$population == populations[i] & data$type == svtype 
    circos.genomicTrack(
      ylim = c(0,1),
      data = sv[svquery, -1],
      track.height = trackheight,
      bg.border = F,
      panel.fun = function(region, value, ...) {
      circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = colormap[i], border = colormap[i])
      }
    )
  }
  text(0, 0, svtype, cex = 2)
}
```

```{r echo = FALSE, message = FALSE, warning = FALSE, fig.align='center', fig.width=12, fig.height=12, out.width= "100%"}
par(mfrow = c(2, 2))
popheight <- 5 / length(populations) / 10

for (SV in c("INV", "DEL", "DUP", "BND")){
  circosplot(sv, fa.sizes, SV, popheight, populations, col.ramp)
}


```

### Per-contig
```{r echo = FALSE, message = FALSE, warning = FALSE}
sv$ystart <- case_when(
  sv$type == "INV" ~ 0.1,
  sv$type == "DUP" ~ 1.1,
  sv$type == "DEL" ~ 2.1,
  sv$type == "BND" ~ 3.1,
)
sv$ystop <- sv$ystart + 0.8
```

```{r echo = FALSE, out.width = '100%', warning=FALSE}
color.palette <- c("DEL"="#5a8c84", "DUP"="#ffd75f", "INV"="#4a9fea","BND"="#df487f")
l <- htmltools::tagList()

for (i in 1:nrow(fa.sizes)) {
  sv.filt <- sv %>% filter(contig == fa.sizes$contig[i])
  sv_stats <- group_by(sv.filt, type) %>% summarise(n = length(type))
  if (nrow(sv_stats) == 0) {
    next
  }
  plt <- sv.filt %>%
    ggplot() +
    geom_rect(
      alpha = 0.7,
      aes(
        ymin = ystart,
        ymax = ystop,
        xmin = position_start,
        xmax = position_end,
        fill = type,
        color = type,
        text = sprintf("Population: %s<br>Type: %s<br>Position: %s-%s<br>barcodes: %s", population, type, position_start, position_end, n_barcodes)
        )
      ) +
    geom_hline(yintercept = 1:3, color = "grey90") +
    scale_color_manual(values = color.palette) +
    scale_fill_manual(values = color.palette) +
    facet_grid(rows = vars(population)) +
    theme_light() +
    theme(
      axis.text.y = element_text(angle=90),
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey80"),
      strip.text = element_text(color = "grey20"),
      legend.position = "none",
    ) +
    xlim(0, fa.sizes$size[i]) +
    scale_y_continuous(
      breaks = c(0.5, 1.5, 2.5, 3.5),
      labels = c("INV", "DUP", "DEL", "BND"),
      limits = c(0,4)
    ) +
    coord_cartesian(xlim = c(0, fa.sizes$size[i] + 1), expand = F) +
    xlab("Position (bp)") +
    labs(fill = "SV type", color = "SV type") +
    ggtitle(fa.sizes$contig[i])
  l[[i]] <- ggplotly(plt, height = 155 * length(populations), tooltip = "text")
}
l
```