---
title: "Sequence Alignment Depth and Coverage"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    toc: no
---

```{r echo = FALSE, message = FALSE, warning = FALSE}
library(ggplot2)
library(dplyr)
library(tidyr)
library(circlize)
library(scales)
```

```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
covfile <- snakemake@input[[1]]
gencov <- read.table(covfile, header = F)
gencov <- gencov[, c(1, 2, 3, 5)]
colnames(gencov) <- c("contig", "start", "end", "depth")
sname <- gsub("\\.\\w?\\w?\\w?.?cov.gz", "", basename(covfile))
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
global_avg <- mean(gencov$depth)
global_sd <- sd(gencov$depth)
zscores <- (gencov$depth - global_avg) / global_sd
gencov$outlier <- zscores >= 3
outliers <- gencov[gencov$outlier, -5]
nonoutliers <- gencov[!(gencov$outlier), -5]
```
#### Sample: `r sname`
#### Aligner: `BWA`

## Summary information {.tabset}
### Averages
The table below shows the global and per-contig average depth and standard 
deviation per 10kbp intervals, **including** intervals whose depth is flagged
an outlier in the data.

```{r echo = FALSE, message = FALSE, warning = FALSE}
contig_avg <- group_by(gencov, contig) %>%
  summarize(average = mean(depth), sdv = sd(depth))
contig_avg <- rbind(data.frame(contig = "global", average = global_avg, sdv = global_sd), contig_avg) %>% 
  mutate(average = round(average, 2), sdv = round(sdv, 2))
DT::datatable(contig_avg, rownames = F, extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE), colnames = c('Contig', 'Average Depth', 'Standard Deviation'))
```

### Filtered Averages
The tabel below shows the global and per-contig average depth and standard 
deviation per 10kbp intervals, **excluding** intervals whose depth is flagged
an outlier in the data, as determined by being greater than 3 standard deviations
above the mean depth. This should be a more accurate representation of read coverage.
```{r echo = FALSE, message = FALSE, warning = FALSE}
global_avg_filt <- mean(nonoutliers$depth)
global_sd_filt <- sd(nonoutliers$depth)
contig_avg_filt <- group_by(nonoutliers, contig) %>%
  summarize(average = mean(depth), sdv = sd(depth))
contig_avg_filt <- rbind(data.frame(contig = "global", average = global_avg_filt, sdv = global_sd_filt), contig_avg_filt) %>% 
  mutate(average = round(average, 2), sdv = round(sdv, 2))

DT::datatable(contig_avg_filt, rownames = F, extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE), colnames = c('Contig', 'Average Depth', 'Standard Deviation'))
```


### Outliers
Below is a table of the intervals with a coverage being greater than 3 standard 
deviations above the mean depth.
```{r echo = FALSE, message = FALSE, warning = FALSE}
DT::datatable(outliers, rownames = F, extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE), colnames = c('Contig', 'Position Start', 'Position End', 'Depth'))
```


## Alignment Depth {.tabset}
Plotted below are (up to) the 30 largest contigs to show sequence alignment depth across
the genome. Each bar represents the alignment depth at a 10kb genomic interval, that is, the
number of reads that had a proper alignment in the 10kb interval. "Proper" refers to a read
not marked as a duplicate or flagged with the SAM `UNMAP`, `SECONDARY`,  or `QCFAIL` flags.
These values are derived by using `samtools bedcov -n`.

### Summary
This is a circular plot summarizing the depth information across the largest contigs.
For clarity, this visualization omits intervals flagged as outliers.
```{r echo = FALSE, message = FALSE, warning = FALSE}
# Find the 30 largest contigs
contigs <- group_by(gencov, contig) %>%
  summarize(size = max(end)) %>%
  arrange(desc(size))

# limit the data to only the 30 largest contigs
if (nrow(contigs) > 30){
    pltheight <- 45
    .contigs <- contigs[1:30, ]
    .gencov <- filter(gencov, contig %in% .contigs)
} else {
    pltheight <- round(1.5 * (nrow(contigs)), digits = 0)
    .gencov <- gencov
    .contigs <- contigs
}

.l <-  nrow(.contigs)
```
```{r echo = FALSE, message = FALSE, warning = FALSE, fig.align='center', out.width= "100%"}
circos.clear()
col_text <- "grey40"
circos.par("track.height" = 0.8, gap.degree = 4, cell.padding = c(0, 0, 0, 0))
circos.initialize(
  factors = .contigs$contig,
  xlim = matrix(c(rep(0, .l), .contigs$size), ncol = 2)
)

# contigs
circos.track(
  ylim=c(0, 1), 
  panel.fun=function(x, y) {
    chr=CELL_META$sector.index
    xlim=CELL_META$xlim
    ylim=CELL_META$ylim
    circos.text(
      mean(xlim),
      mean(ylim),
      chr,
      cex=0.5,
      col=col_text,
      facing="bending.inside",
      niceFacing = TRUE
      )
    },
  bg.col = "grey90",
  bg.border = F,
  track.height = 0.06
)

# x axis
brk <- seq(0, 10, 0.5) * 10^7
circos.track(
  track.index = get.current.track.index(),
  panel.fun = function(x, y) {
    circos.axis(
      h = "top",
      major.at = brk,
      labels = round(brk / 10^6, 1),
      labels.cex = 0.4,
      col = col_text,
      labels.col = col_text,
      lwd = 0.7,
      labels.facing = "clockwise"
    )
  },
  bg.border = F
)

maxaln <- round(max(nonoutliers$depth), digits = -1)
cov.yaxis <- c(0, round(maxaln * 1/3, digits = -1), round(maxaln * 2/3, digits = -1), maxaln)

# coverage
circos.genomicTrack(
  data = nonoutliers,
  panel.fun = function(region, value, ...) {
    circos.genomicLines(region, value, type = "l", col = "#3f3f3f", lwd = 0.8)
    circos.segments(x0=0, x1=max(.contigs$size), y0=min(nonoutliers$depth)-100, y1=min(nonoutliers$depth) - 100, lwd=0.8, lty="11", col="grey90")
  },
  track.height=0.5,
  bg.border = F
)
circos.yaxis(at=cov.yaxis, labels.cex=0.4, lwd=0.8, labels.col=col_text, col="grey30")

```

### Per-Contig
Plotted below is the same depth information, but shown in better detail per 
contig for easier visual assessment. Note that regions of extremely high depth
are flagged as outliers and for visual clarity, their values have been converted
to the global mean depth (`r round(global_avg_filt,0)`) after removing outliers. Investigate
the table above for more information about these regions.

```{r echo = FALSE, message = FALSE, warning = FALSE}
# Cap the outliers as the mean
.gencov$depth[.gencov$outlier] <- global_avg_filt
```


```{r echo = FALSE, message = FALSE, warning = FALSE, fig.height=pltheight, out.width = "100%"}
  ggplot(.gencov, aes(xmin = start, xmax = end, ymin = 0, ymax = depth)) +
    geom_rect(alpha = 0.7, aes(color = outlier, fill = outlier)) +
    scale_color_manual(values = c("FALSE" = "#3f3f3f", "TRUE" =  "#cb6843")) +
    scale_fill_manual(values = c("FALSE" = "#3f3f3f", "TRUE" =  "#cb6843")) +
    theme_minimal() +
    facet_wrap(~contig, ncol = 2, scales = "free_x") +
    labs(x = "position (bp)", y = "depth") +
    scale_y_continuous(labels = scales::comma) +
    theme(
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      legend.position = "top"
      ) +
    guides(x = "none") -> p

print(p)
```