---
title: "Leviathan Variant Calling Per-Sample Summary"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    df_print: kable
    toc: no
---

Below is a series of tables and plots detailing the structural variants
identified by `LEVIATHAN`. The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
```

```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(magrittr)
library(DT)
library(plotly)
library(circlize)
library(ComplexHeatmap)
```

```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r contigSizes, echo = FALSE}
bcf <- snakemake@input[["bcf"]]
#bcf <- "/home/pdimens/Omega/Cornell/dmel/leviathan.bcf"
hd <- system2("bcftools", args =  paste("head", bcf), stdout = T)
fa.sizes <- hd[grep("contig=", hd)]
fa.sizes <- as.data.frame(do.call(rbind, strsplit(fa.sizes, ",length="))) %>% arrange(desc(V2))
fa.sizes$V1 <- gsub("##contig=<ID=", "", fa.sizes$V1)
fa.sizes$V2 <- as.numeric(gsub(">", "", fa.sizes$V2))
colnames(fa.sizes) <- c("contig", "size")
# limit the data to only the 30 largest contigs
if (nrow(fa.sizes) > 30){
  fa.sizes <- fa.sizes[1:30, ]
}
```


## Various Stats {.tabset}
### all variants
This table details all the variants LEVIATHAN detected and the quantity/quality of evidence for those detections.
```{r echo = FALSE, warnings = FALSE, message = FALSE}
statsfile <- snakemake@input[["statsfile"]]
#statsfile <- "/home/pdimens/Omega/Cornell/dmel/leviathan.stats"
sv <- read.table(statsfile, header = T)
if (nrow(sv) == 0) {
  print("No variants in the file")
  q()
}
DT::datatable(sv, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

### summary info
This table shows the overall statistics for detected structural variants.
```{r echo = FALSE, message = FALSE}
sv$length <- gsub("\\.", "1", sv$length) %>%  as.numeric()

grpstats <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n(), total_bp  = sum(length))

grpstats %>%
  group_by(type) %>%
  summarise(
    total = sum(count),
    avg_per_contig = mean(count),
    sd = sd(count)
)
```

### per-contig info
This table details each type of structural variant for every contig.
```{r echo = FALSE, message = FALSE, warning = FALSE}
grpstats <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n(), total_bp  = sum(length))

DT::datatable(grpstats, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

---

## SV size and location {.tabset}
### Summary
Below is a circular plot to visualize the distribution of SV across (up to) 30 of the largest contigs. This should help you assess
the presence/absence of expected variants. For more detailed interactive plots, see the **Per-contig** tab. You may click on the plot
to expand it.
```{r echo = FALSE, message = FALSE, warning = FALSE, fig.align='center', out.width= "100%"}
circos.clear()
col_text <- "grey40"
circos.par("track.height"=0.8, gap.degree=4, cell.padding=c(0, 0, 0, 0))
circos.initialize(
  factors = fa.sizes$contig, 
  xlim = matrix(c(rep(0, nrow(fa.sizes)), fa.sizes$size), ncol=2)
)

# contigs
circos.track(
  ylim=c(0, 1), 
  panel.fun=function(x, y) {
    chr=CELL_META$sector.index
    xlim=CELL_META$xlim
    ylim=CELL_META$ylim
    circos.text(
      mean(xlim),
      mean(ylim), 
      chr, 
      cex=0.5, 
      col=col_text, 
      facing="bending.inside", 
      niceFacing=TRUE
      )
    }, 
  bg.col="grey90", 
  bg.border=F, 
  track.height=0.06
)

# x axis
brk <- seq(0, 10, 0.5)*10^7
circos.track(
  track.index = get.current.track.index(), 
  panel.fun=function(x, y) {
    circos.axis(
      h="top", 
      major.at=brk, 
      labels=round(brk/10^6, 1), 
      labels.cex=0.4, 
      col=col_text, 
      labels.col=col_text, 
      lwd=0.7, 
      labels.facing="clockwise"
    )
  }, 
  bg.border=F
)

# INV
circos.genomicTrack(
  ylim = c(0,1),
  data = sv[sv$type == "INV", -1], 
  track.height = 0.2,
  bg.border = F,
  panel.fun=function(region, value, ...) {
    circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = "#1E88E5", border = "#1E88E5")
  }
)


circos.genomicTrack(
  ylim = c(0,1),
  data = sv[sv$type == "DEL", -1], 
  track.height = 0.2,
  bg.border = F,
  panel.fun=function(region, value, ...) {
    circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = "#004D40", border = "#004D40")
  }
)

circos.genomicTrack(
  ylim = c(0,1),
  data = sv[sv$type == "DUP", -1],
  track.height = 0.2,
  bg.border = F,
  panel.fun=function(region, value, ...) {
    circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = "#FFC107", border =  "#FFC107")
  }
)

circos.genomicTrack(
  ylim = c(0,1),
  data = sv[sv$type == "BND", -1],
  track.height = 0.2,
  bg.border = F,
  panel.fun=function(region, value, ...) {
    circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = "#D81B60", border = "#D81B60")
  }
)

lgd <- Legend(
  at = c("Inversion", "Deletion", "Duplication", "Breakend"),
  type = "rect", 
    legend_gp = gpar(col = c("#1E88E5", "#004D40", "#FFC107", "#D81B60")),
    title_position = "topleft",
    title = "Variant Types"
)

draw(lgd, x = unit(0.9, "npc"), just = c("right", "bottom"))
```

### Per-contig
Below are a series of plots to help you assess what structural variants were detected by LEVIATHAN. These plots are interactive,
allowing you to hover over a variant to provide additional information, including the genomic interval in which it occurs and the
number of haplotag barcodes supporting the variant.
```{r echo = FALSE, out.width = '100%', warning=FALSE}
l <- htmltools::tagList()
for (i in 1:nrow(fa.sizes)) {
  sv.filt <- sv %>% filter(contig == fa.sizes$contig[i])
  sv_stats <- group_by(sv.filt, type) %>%  summarise(n = length(type))
  if (nrow(sv_stats) == 0) {
    next
  }
  plt <- sv.filt %>%
    ggplot() +
    geom_rect(
      alpha = 0.7,
      aes(
        xmin = position_start,
        xmax = position_end,
        ymin = 0,
        ymax = 1,
        fill = type,
        color = type,
        text = sprintf("Type: %s<br>Position: %s-%s<br>barcodes: %s", type, position_start, position_end, n_barcodes)
      )
    ) +
    theme_light() +
    facet_grid(rows = vars(type)) +
    theme(
      axis.text.y=element_blank(),
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey80"),
      strip.text = element_text(color = "grey20"),
      legend.position = "none"
    ) +
    scale_color_manual(values=c("DEL"="#004D40", "DUP"="#FFC107", "INV"="#1E88E5","BND"="#D81B60")) +
    scale_fill_manual(values=c("DEL"="#004D40", "DUP"="#FFC107", "INV"="#1E88E5","BND"="#D81B60")) +
    xlim(1, fa.sizes$size[i]) +
    coord_cartesian(xlim = c(0, fa.sizes$size[i] + 1), expand = F) +
    xlab("Position (bp)") +
    labs(fill = "SV type", color = "SV type") +
    ggtitle(i)
  l[[i]] <- ggplotly(plt, tooltip = "text")
}
l
```
