---
title: "Sequence Alignment Depth and Coverage"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    toc: no
---

```{r echo = FALSE, message = FALSE, warning = FALSE}
library(ggplot2)
library(dplyr)
library(tidyr)
library(circlize)
library(scales)
```

```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
covfile <- snakemake@input[[1]]
gencov <- read.table(covfile, header = F)
gencov <- gencov[, c(1, 2, 3, 6, 7)]
colnames(gencov) <- c("contig", "start", "end", "Barcode", "noBarcode")
gencov$noBarcode <- gencov$noBarcode * -1
sname <- gsub("\\.\\w?\\w?\\w?.?cov.gz", "", basename(covfile))
```
#### Sample: `r sname`



```{r echo = FALSE, message = FALSE, warning = FALSE}
bc_avg <- mean(gencov$Barcode)
bc_sd <- sd(gencov$Barcode)
nobc_avg <- mean(abs(gencov$noBarcode))
nobc_sd <- sd(abs(gencov$noBarcode))
all_depth <- gencov$Barcode + abs(gencov$noBarcode)
gencov$all <- all_depth
global_avg <- mean(all_depth)
global_sd <- sd(all_depth)

zscores <- (all_depth - global_avg) / bc_sd
gencov$outlier <- zscores >= 3
outliers <- gencov[gencov$outlier, -6]
nonoutliers <- gencov[!(gencov$outlier), -6]
```

## Summary information {.tabset}
### Averages
The table below shows the global and per-contig average depth and standard 
deviation per 10kbp intervals, **including** intervals whose depth is flagged
an outlier in the data.
```{r echo = FALSE, message = FALSE, warning = FALSE}
# pivot longer for generalized plotting
Lgencov <- pivot_longer(gencov[,-7], -(1:3), names_to = "alignment", values_to = "depth")
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
contig_avg <- group_by(Lgencov, contig, alignment) %>%
  summarize(average = round(mean(abs(depth)),2), sdv = round(sd(abs(depth)),2))

contig_avg <- pivot_wider(data = contig_avg, names_from = alignment,values_from = c(average,sdv))[, c(1,3,6,2,5,4,7)]

DT::datatable(contig_avg, rownames = F, extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE), colnames = c('Average All', 'StDev All', 'Average Barcode', 'StDev Barcode', 'Average noBarcode', 'StDev noBarcode'))
```

### Filtered Averages
The tabel below shows the global and per-contig average depth and standard 
deviation per 10kbp intervals, **excluding** intervals whose depth is flagged
an outlier in the data, as determined by being greater than 3 standard deviations
above the mean depth. This should be a more accurate representation of read coverage.
```{r echo = FALSE, message = FALSE, warning = FALSE}
nonoutliers$all <- nonoutliers$Barcode + abs(nonoutliers$noBarcode)
bc_avg <- mean(nonoutliers$Barcode)
bc_sd <- sd(nonoutliers$Barcode)
nobc_avg <- mean(abs(nonoutliers$noBarcode))
nobc_sd <- sd(abs(nonoutliers$noBarcode))
all_depth <- nonoutliers$all
global_avg <- mean(all_depth)
global_sd <- sd(all_depth)

filt_df <- data.frame(
  "contig" = "global",
  "average_all" = round(global_avg,2),
  "sdv_all" = round(global_sd,2),
  "average_Barcode" = round(bc_avg,2),
  "sdv_Barcode" = round(bc_sd,2),
  "average_noBarcode" = round(nobc_avg,2),
  "sdv_noBarcode" = round(nobc_sd,2)
  )
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
Lnonoutliers <- pivot_longer(nonoutliers[,-6], -(1:3), names_to = "alignment", values_to = "depth")


contig_avg_filt <- group_by(Lnonoutliers, contig, alignment) %>%
  summarize(average = round(mean(abs(depth)),2), sdv = round(sd(abs(depth)),2))

contig_avg_filt <- pivot_wider(data = contig_avg_filt, names_from = alignment,values_from = c(average,sdv))[, c(1,3,6,2,5,4,7)]

contig_avg_filt <- rbind(filt_df, contig_avg_filt)

DT::datatable(contig_avg_filt, rownames = F, extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE), colnames = c('Average All', 'StDev All', 'Average Barcode', 'StDev Barcode', 'Average noBarcode', 'StDev noBarcode'))
```


### Outliers
Below is a table of the intervals with a coverage being greater than 3 standard 
deviations above the mean depth. `Barcode Depth` refers to the alignment depth of sequences with a valid
haplotag barcode, `noBarcode Depth` refers to the alignment depth of sequences without a valid haplotag
barcode, and `Total Depth` is the sum of `Barcode Depth` and `noBarcode Depth`.
```{r echo = FALSE, message = FALSE, warning = FALSE}
outliers$noBarcode <- abs(outliers$noBarcode)
outliers$all <- outliers$Barcode + abs(outliers$noBarcode)
DT::datatable(outliers[,-6], rownames = F, extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE), colnames = c('Contig', 'Position Start', 'Position End', 'Barcode Depth', 'noBarcode Depth', 'Total Depth'))
```


## Alignment Depth {.tabset}
Plotted below are (up to) the 30 largest contigs to show sequence alignment depth across
the genome. Each bar represents the alignment depth at a 10kb genomic interval, that is, the
number of reads that had a proper alignment in the 10kb interval. "Proper" refers to a read
not marked as a duplicate or flagged with the SAM `UNMAP`, `SECONDARY`,  or `QCFAIL` flags.
These values are derived by using `samtools bedcov -n`.

### Summary
This is a circular plot summarize the information across the largest contigs, with black showing
alignment depth of the barcoded sequences. The reddish color shows the alignment depth
of sequences with an invalid barcode. For clarity, this visualization omits intervals
flagged as outliers.

```{r echo = FALSE, message = FALSE, warning = FALSE}
# Find the 30 largest contigs
contigs <- group_by(Lnonoutliers, contig) %>%
  summarize(size = max(end)) %>%
  arrange(desc(size))

# limit the data to only the 30 largest contigs
if (nrow(contigs) > 30) {
    pltheight <- 45
    .contigs <- contigs[1:30, ]
    .gencov <- filter(Lnonoutliers, contig %in% .contigs$contig) %>% 
      filter(alignment != "all")
} else {
    pltheight <- round(1.5 * (nrow(contigs)), digits = 0)
    .gencov <- Lnonoutliers %>% filter(alignment != "all")
    .contigs <- contigs
}

.l <-  nrow(.contigs)
```
```{r echo = FALSE, message = FALSE, warning = FALSE, fig.align='center', out.width= "100%"}
nonoutliers$noBarcode <- nonoutliers$noBarcode * -1

circos.clear()
col_text <- "grey40"
circos.par("track.height" = 0.8, gap.degree = 4, cell.padding = c(0, 0, 0, 0))
circos.initialize(
  factors = .contigs$contig,
  xlim = matrix(c(rep(0, .l), .contigs$size), ncol = 2)
)

# contigs
circos.track(
  ylim = c(0, 1),
  panel.fun = function(x, y) {
    chr = CELL_META$sector.index
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    circos.text(
      mean(xlim),
      mean(ylim),
      chr,
      cex = 0.5,
      col = col_text,
      facing = "bending.inside",
      niceFacing = TRUE
      )
    },
  bg.col = "grey90",
  bg.border = F,
  track.height = 0.06
)

# x axis
brk <- seq(0, 10, 0.5) * 10^7
circos.track(
  track.index = get.current.track.index(),
  panel.fun = function(x, y) {
    circos.axis(
      h = "top",
      major.at = brk,
      labels = round(brk / 10^6, 1),
      labels.cex = 0.4,
      col = col_text,
      labels.col = col_text,
      lwd = 0.7,
      labels.facing = "clockwise"
    )
  },
  bg.border = F
)

maxaln <- max(abs(.gencov$depth))
cov.yaxis <- c(0, round(maxaln * 1/3, digits = -1), round(maxaln * 2/3, digits = -1), maxaln)

# depth
circos.genomicTrack(
  data=nonoutliers, 
  panel.fun=function(region, value, ...) {
    circos.genomicLines(region, value, type="l", col="#3f3f3f", lwd=0.8)
    circos.genomicLines(region, value, type="l", col="#cb6843", lwd=0.8, numeric.column = 2)
  }, 
  track.height=0.5,
  bg.border=F
)
circos.yaxis(at=cov.yaxis, labels.cex=0.4, lwd=0.8, labels.col=col_text, col="grey30")

```

### Per-Contig
Plotted below is the same depth information, but shown in better detail per 
contig for easier visual assessment. Note that regions of extremely high depth
flagged as outliers appear in blue and for visual clarity, their values have been converted
to the global mean depth recalculated after removing outliers (`r round(global_avg , 0)`). Investigate
the table above for more information about these regions. The non-barcoded alignment depths are multiplied
by -1 to be visible on the opposite side of the X-axis.

```{r echo = FALSE, message = FALSE, warning = FALSE}
# pivot longer for generalized plotting
gencov <- pivot_longer(gencov[,-6],  4:5, names_to = "alignment", values_to = "depth")
gencov <- filter(gencov, contig %in% .contigs$contig)
gencov$depth[gencov$outlier] <- global_avg

outliers <- pivot_longer(outliers, 4:5, names_to = "alignment", values_to = "depth")
```

```{r echo = FALSE, message = FALSE, warning = FALSE, fig.height=pltheight, out.width = "100%"}
ggplot(gencov[!(gencov$outlier),], aes(xmin = start, xmax = end, ymin = 0, ymax = depth, color = alignment, fill = alignment)) +
  geom_rect(alpha = 0.7) +
  geom_rect(data = outliers, color = "dodgerblue", fill = "dodgerblue", aes(xmin = start, xmax = end, ymin = 0, ymax = global_avg)) +
  #geom_point(data = outliers, x = (outliers$end - outliers$start)/2, y = 0, color = "purple") +
  theme_minimal() +
  facet_wrap(~contig, ncol = 2, scales = "free_x") +
  scale_color_manual(values = c("#3f3f3f", "#cb6843")) +
  scale_fill_manual(values = c("#3f3f3f", "#cb6843")) +
  scale_y_continuous(labels = scales::comma) +
  labs(x = "position (bp)", y = "depth", title = "Alignment Depth") +
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    legend.position = "top"
    ) +
  guides(x = "none") -> p

print(p)
```