---
title: "NAIBR Structural Variants Summary"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    df_print: kable
    toc: no
---

```{r echo = FALSE, warnings = FALSE, message = FALSE}
infiles <- snakemake@input[["bedpe"]]
fai <- snakemake@input[["fai"]]
#infiles <- c("/home/pdimens/ZS10.ema.bedpe", "/home/pdimens/ZS10.bwa.bedpe")
#fai <- "/home/pdimens/assembly.fai"
```

Below is a series of tables and plots detailing the structural variants
identified by [NAIBR](https://github.com/pontushojer/NAIBR) [(publication)](https://doi.org/10.1093/bioinformatics/btx712).
The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
data.frame(
  "Name" = c("Inversion", "Duplication", "Deletion"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA"
    )
  )
```
```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(magrittr)
library(tidyr)
library(DT)
library(plotly)
library(circlize)
library(viridisLite)
library(tools)
library(ComplexHeatmap)
```
```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r echo = FALSE, warnings = FALSE, message = FALSE}
readvariants <- function(x){
  .x <- read.table(x, header = T)
  popname <- data.frame("Population" = rep(gsub(".bedpe", "", basename(x)), nrow(.x)))
  y <- cbind(popname, .x)
  return(y)
}

cleanbedpe <- function(x){
  y <- x[x$Chr1 == x$Chr2, c(-4, -11)]
  names(y)[2] <- "Contig"
  y$Score <- round(y$Score, digits = 3)
  y$Length <- abs(y$Break2 - y$Break1)
  return(y)
}


chimeric <- function(x){
  y <- x[x$Chr1!=x$Chr2, -11]
  y$Score <- round(y$Score, digits = 3)
  return(y)
}
```

```{r echo = FALSE, warnings = FALSE, message = FALSE}
fa.sizes <- read.table(fai, header = F) %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")
# limit the data to only the 30 largest contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]
```

```{r echo = FALSE, results='asis'}
sv <- do.call(rbind, lapply(infiles, function(x) readvariants(x)))

if (nrow(sv) == 0) {
  cat("There are no variants detected in files:\n")
  cat(paste0("- `", infiles, "`"), sep = "\n")
  knitr::knit_exit()
}
```

## Various information {.tabset}
### summary info
This table shows the overall counts, average per contig, and standard deviation per contig for detected structural variants across populations.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
sv_clean <- cleanbedpe(sv)
genstats <- sv_clean %>%
  group_by(Population, Contig, SV) %>%
  summarise(tot = n()) %>%
  group_by(Population, SV) %>%
  summarise(count = sum(tot), avg = round(mean(tot, na.rm = T),2 ), sd = round(sd(tot, na.rm = T), 2)) %>% 
  pivot_wider(names_from = SV, values_from = c(count, avg, sd), names_glue = "{SV}.{.value}", names_vary = "slowest")

nicecols <- toTitleCase(gsub("\\."," ", names(genstats)))
DT::datatable(
  genstats,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = nicecols
)
```

### all variants
This table details the variants detected by NAIBR that appear on a single contig/chromosome and passed the programs internal filtering.
```{r message = FALSE, echo = FALSE}
DT::datatable(
  sv_clean,
  rownames = F,
  filter = "top",
  extensions = 'Buttons', 
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE)
)
```
### chimeric variants

This table shows any structural variants whose breakpoints span multiple contigs and may require further assessment. These variants are not shown in the plots below.

```{r echo = FALSE, warnings = FALSE, message = FALSE}
DT::datatable(
  chimeric(sv),
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE)
)

```

## Per-contig information {.tabset}
### counts per contig
This table details counts each type of structural variant for every contig for every population.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grpcounts <- sv_clean %>%
  group_by(Population, Contig, SV) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = Population, values_from = count)

DT::datatable(
  grpcounts,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE)
)
```

### bp per contig
This table details the lengths (in bp) of structural variants for every contig for every population.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grplens <- sv_clean %>% 
  group_by(Population, Contig, SV) %>%
  summarise(bp = sum(Length)) %>%
  pivot_wider(names_from = Population, values_from = bp)

DT::datatable(
  grplens,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE)
)
```

---

## SV size and location {.tabset}
The plots below are a visual representation of each SV where it was detected.
Contigs that are absent did not have any SV's detected on them. Within each 
contig, each population is plotted as a row to help visualize shared
variants between populations. Plotting only the first 30 largest contigs.

### Summary
```{r echo = FALSE, message = FALSE, warning = FALSE}
circosplot <- function(data, contigdata, svtype, trackheight, poplist, colormap){
  circos.clear()
  col_text <- "grey40"
  circos.par("track.height" = 0.8, gap.degree = 2, cell.padding = c(0, 0, 0, 0))
  circos.initialize(
    factors = contigdata$contig,
    xlim = matrix(c(rep(0, nrow(contigdata)), contigdata$size), ncol = 2)
  )

  # contigs
  circos.track(
    ylim = c(0, 1),
    bg.col = "grey90",
    bg.border = F,
    track.height = 0.06,
    panel.fun = function(x, y) {
      chr = CELL_META$sector.index
      xlim = CELL_META$xlim
      ylim = CELL_META$ylim
      circos.text(
        mean(xlim),
        mean(ylim),
        chr,
        cex = 0.6,
        col = col_text,
        facing="bending.inside",
        niceFacing = T
      )
    }
  )

  # x axis
  brk <- seq(0, 10, 0.5) * 10^7
  circos.track(
    ylim = c(0,1),
    track.index = get.current.track.index(),
    bg.border = F,
    panel.fun = function(x, y) {
      circos.axis(
        h = "top",
        major.at = brk,
        labels = round(brk/10^6, 1),
        labels.cex = 0.6,
        col = col_text,
        labels.col = col_text,
        lwd = 0.7,
        labels.facing = "clockwise"
      )
    }
  )

  # plot the track
  for(i in 1:length(poplist)){
    svquery = data$Population == populations[i] & data$SV == svtype
    circos.genomicTrack(
      ylim = c(0,1),
      data = data[svquery, -1],
      track.height = trackheight,
      bg.border = F,
      panel.fun = function(region, value, ...) {
      circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = colormap[i], border = colormap[i])
      }
    )
  }
  text(0, 0, svtype, cex = 2)
}
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
# add legend to an empty plot
# used in the next chunk to add the legend to the last cell in the 2x2 grid.
draw_legend <- function(contigdata, poplist, colormap){
  circos.clear()
  col_text <- "grey40"
  circos.par("track.height" = 0.8, gap.degree = 2, cell.padding = c(0, 0, 0, 0))
  circos.initialize(
    factors = contigdata$contig,
    xlim = matrix(c(rep(0, nrow(contigdata)), contigdata$size), ncol = 2)
  )
  lgd <- Legend(
      at = poplist,
      type = "grid",
      legend_gp = gpar(fill = colormap),
      labels_gp = gpar(fontsize = 16),
      title_position = "topleft",
      title = "Populations",
      title_gp = gpar(fontsize = 20)
      
    )
    draw(lgd, x = unit(250, "mm"), y = unit(140, "mm"))
}
```

```{r echo = FALSE, message = FALSE, warning = FALSE, fig.align='center', fig.width=13, fig.height=13, out.width= "100%"}
par(mfrow = c(2, 2))
populations <- unique(sv$Population)
popheight <- .5 / length(populations)
col.ramp <- viridis(length(populations))

for (SV in c("inversion", "deletion", "duplication")){
  circosplot(sv_clean, fa.sizes, SV, popheight, populations, col.ramp)
}
draw_legend(fa.sizes, populations, col.ramp)

```

### Per-contig
```{r echo = FALSE, message = FALSE, warning = FALSE}
sv_clean$ystart <- case_when(
  sv_clean$SV == "inversion" ~ 0.1,
  sv_clean$SV == "duplication" ~ 1.1,
  sv_clean$SV == "deletion" ~ 2.1
)
sv_clean$ystop <- sv_clean$ystart + 0.8

chrheight <- paste0(1.5 * length(populations), "in")

```

```{r echo = FALSE, out.width = '100%', warning=FALSE, out.height=chrheight}
color.palette <- c("deletion"="#5a8c84", "duplication"="#ffd75f", "inversion"="#4a9fea")
l <- htmltools::tagList()

for (i in 1:nrow(fa.sizes)) {
  sv.filt <- sv_clean %>% filter(Contig == fa.sizes$contig[i])
  sv_stats <- group_by(sv.filt, SV) %>% summarise(n = length(SV))
  if (nrow(sv_stats) == 0) {
    next
  }
  plt <- sv.filt %>%
    ggplot() +
    geom_rect(
      alpha = 0.7,
      aes(
        ymin = ystart,
        ymax = ystop,
        xmin = Break1,
        xmax = Break2,
        fill = SV,
        color = SV,
        text = sprintf("Population: %s<br>Type: %s<br>Position: %s-%s<br>barcodes: %s", Population, SV, Break1, Break2, SplitMolecules)
        )
      ) +
    geom_hline(yintercept = 1:2, color = "grey90") +
    scale_color_manual(values = color.palette) +
    scale_fill_manual(values = color.palette) +
    facet_grid(rows = vars(Population)) +
    theme_light() +
    theme(
      axis.text.y = element_text(angle=90),
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey80"),
      strip.text = element_text(color = "grey20"),
      legend.position = "none",
    ) +
    xlim(0, fa.sizes$size[i]) +
    scale_y_continuous(
      breaks = c(0.5, 1.5, 2.5),
      labels = c("Inv", "Dup", "Del"),
      limits = c(0,3)
    ) +
    coord_cartesian(xlim = c(0, fa.sizes$size[i] + 1), expand = F) +
    xlab("Position (bp)") +
    labs(fill = "SV type", color = "SV type") +
    ggtitle(fa.sizes$contig[i])
  l[[i]] <- ggplotly(plt, tooltip = "text")
}
l
```