---
title: "Leviathan Variant Calling Per-Sample Summary"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    df_print: kable
    toc: no
---

Below is a series of tables and plots detailing the structural variants
identified by `LEVIATHAN`. The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
```

```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(magrittr)
library(DT)
library(plotly)
```

```{r contigSizes, echo = FALSE}
hd <- system2("bcftools", args =  paste("head", snakemake@input[["bcf"]]), stdout = T)
fa.sizes <- hd[grep("contig=", hd)]
fa.sizes <- as.data.frame(do.call(rbind, strsplit(fa.sizes, ",length="))) %>% arrange(desc(V2))
fa.sizes$V1 <- gsub("##contig=<ID=", "", fa.sizes$V1)
fa.sizes$V2 <- as.numeric(gsub(">", "", fa.sizes$V2))
colnames(fa.sizes) <- c("contig", "size")
# limit the data to only the 30 largest contigs
if (nrow(fa.sizes) > 30){
  fa.sizes <- fa.sizes[1:30, ]
}
```


## Various Stats {.tabset}
### all variants
This table details all the variants LEVIATHAN detected and the quantity/quality of evidence for those detections.
```{r echo = FALSE, warnings = FALSE, message = FALSE}
sv <- read.table(snakemake@input[["statsfile"]], header = T)
if (nrow(sv) == 0) {
  print("No variants in the file")
  q()
}
DT::datatable(sv, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

### summary info
This table shows the overall statistics for detected structural variants.
```{r echo = FALSE, message = FALSE}
sv$length <- gsub("\\.", "1", sv$length) %>%  as.numeric()

grpstats <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n(), total_bp  = sum(length))

grpstats %>%
  group_by(type) %>%
  summarise(
    total = sum(count),
    avg_per_contig = mean(count),
    sd = sd(count)
)
```

### per-contig info
This table details each type of structural variant for every contig.
```{r echo = FALSE, message = FALSE, warning = FALSE}
grpstats <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n(), total_bp  = sum(length))

DT::datatable(grpstats, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

---

## SV size and location
The plots below are a visual representation of each SV where it was detected.
Contigs that are absent did not have any SV's detected on them. Plotting the first 30 largest contigs.

```{r echo = FALSE, out.width = '100%', warning=FALSE}
l <- htmltools::tagList()
for (i in 1:nrow(fa.sizes)) {
  sv.filt <- sv %>% filter(contig == fa.sizes$contig[i])
  sv_stats <- group_by(sv.filt, type) %>%  summarise(n = length(type))
  if (nrow(sv_stats) == 0) {
    next
  }
  plt <- sv.filt %>%
    ggplot() +
    geom_rect(
      alpha = 0.7,
      aes(
        xmin = position_start,
        xmax = position_end,
        ymin = 0,
        ymax = 1,
        fill = type,
        color = type,
        text = sprintf("Type: %s<br>Position: %s-%s<br>barcodes: %s", type, position_start, position_end, n_barcodes)
      )
    ) +
    theme_light() +
    facet_grid(rows = vars(type)) +
    theme(
      axis.text.y=element_blank(),
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey80"),
      strip.text = element_text(color = "grey20"),
      legend.position = "none"
    ) +
    scale_color_manual(values=c("DEL"="#004D40", "DUP"="#FFC107", "INV"="#1E88E5","BND"="#D81B60")) +
    scale_fill_manual(values=c("DEL"="#004D40", "DUP"="#FFC107", "INV"="#1E88E5","BND"="#D81B60")) +
    xlim(1, fa.sizes$size[i]) +
    coord_cartesian(xlim = c(0, fa.sizes$size[i] + 1), expand = F) +
    xlab("Position (bp)") +
    labs(fill = "SV type", color = "SV type") +
    ggtitle(i)
  l[[i]] <- ggplotly(plt, tooltip = "text")
}
l
```