---
title: "Haplotype Phasing Report"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    df_print: kable
    toc: no
---

```{r echo= FALSE, message = FALSE, warning = FALSE}
library(ggplot2)
library(DT)
library(dplyr)
library(magrittr)
library(scales)
library(ggridges)
```

```{r echo= FALSE, message = FALSE, warning = FALSE}
## File
#file <- "~/blocks.file.gz"
file <- snakemake@input[[1]]
df <- read.table(file, header = T)
df$pos_end <- df$pos_start + df$block_length
df <- df[, c(1,2,3,4,6,5)]
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
# Find the 30 largest contigs
contigs <- group_by(df, contig) %>%
  summarize(size = max(pos_end)) %>%
  arrange(desc(size))
# limit the data to only the 30 largest contigs
if (nrow(contigs) > 30){
    .contigs <- contigs[1:30, ]$contig
} else {
    .contigs <- contigs$contig
}
pltheight <- round(2.1 * (length(.contigs)), digits = 0)
ridgeheight <- 1 + round(0.7 * (length(.contigs)), digits = 0)
```

## Haplotyping details {.tabset}
The data presented here are the results of phasing genotypes into haplotypes using
[HapCut2](https://github.com/vibansal/HapCUT2). The information is derived from `r file`,
which summarized information across all samples using the `.blocks` files HapCut2
generates. The VCF files HapCut2 also generates were not used as they lack
some of the information in the `.blocks` files that were collated in this report.

### overall stats
The table below summarizes the outcome of haplotype phasing across all samples and contigs.

```{r echo= FALSE, message = FALSE, warning = FALSE}
overall <- df %>% summarise(
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
)
names(overall) <- c("Average SNPs per Haplotype", "Median SNPs per Haplotype", "Mean Haplotype Size", "Median Haplotype Length", "Largest Haplotype")
overall

```
The plot below shows the distribution of haplotype length (in base pairs). Phasing typically 
results in extreme right-tails in these distributions, making regular visualization
difficult to plot meaningfully. The plot presented below has **log-scaled lengths**
to collapse the right tail for better readability. 

```{r warning=FALSE, message=FALSE, echo= FALSE, out.width="100%", fig.height=2.7}
ggplot(df, aes(block_length, after_stat(count))) + 
  geom_density(color = "#859f9a", fill = "#bfe4dc") +
  theme_minimal() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.minor.x = element_blank()
    ) +
   scale_y_continuous(labels = label_comma()) +
   scale_x_log10(
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))
  ) +
  xlab("Haplotype Length (log scale)") +
  ylab("Count") +
  labs(title = "Distribution of Overall Haplotype Lengths")
```

### stats per contig
The table below provides details on the outcome of haplotype phasing on a per-contig basis.
```{r echo= FALSE, message = FALSE, warning = FALSE}
percontig <- df %>% group_by(contig) %>% summarise(
    n_haplo = round(length(n_snp)),
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
  )

DT::datatable(
  percontig,
  rownames = F,
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = c("Contig", "Total Haplotypes", "Mean SNPs", "Median SNPs", "Mean Haplotype Length", "Median Haplotype Length", "Largest Haplotype")
  )
```

The plot below shows the distribution of haplotype length (in base pairs) for 
each contig, up to 30 of the largest contigs, the size of whom is inferred/assumed
from the end position of the last haplotype. Phasing typically results in extreme
right-tails in these distributions, making regular visualization
difficult to plot meaningfully. The plot presented below has **log-scaled lengths**
to collapse the right tails for better readability. 

```{r warning=FALSE, message=FALSE, echo= FALSE, fig.height=pltheight, out.width="100%"}
ggplot(df, aes(block_length, after_stat(count))) + 
  geom_density(color = "#af9caf", fill = "#DBC3DB") +
  theme_light() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.minor.x = element_blank(),
    strip.background =element_rect(fill="grey90", color = "grey70"),
    strip.text = element_text(colour = 'grey40')
    ) +
  scale_y_continuous(labels = label_comma()) +
  scale_x_log10(
     breaks = trans_breaks("log10", function(x) 10^x),
     labels = trans_format("log10", math_format(10^.x))
  ) +
  facet_wrap(~contig, ncol = 1, scales = "free") +
  xlab("Haplotype Length (log scale)") +
  ylab("Count") +
  labs(title = "Distribution of Haplotype Lengths Per Contig")
```

### stats per sample
Haplotype phasing occurs per-sample and the table below provides details on the outcome of haplotype phasing for each sample.

```{r echo= FALSE, message = FALSE, warning = FALSE}
persample <- df %>% group_by(sample) %>% summarise(
    n_haplo = sum(length(n_snp)),
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
)

DT::datatable(
  persample,
  rownames = F,
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = c("Contig", "Haplotypes", "Mean SNPs", "Median SNPs", "Mean Haplotype Length", "Median Haplotype Length", "Largest Haplotype")
)
```
The ridgeline plot below shows the distribution of haplotype lengths (in base pairs) for each 
sample. While no Y-axis for these counts are provided, this plot is intended to be more of
a visual reference of the relative distributions of haplotype lengths among samples.
Phasing typically results in extreme right-tails in these distributions, making regular 
visualization difficult to plot meaningfully. The plot presented below has 
**log-scaled lengths** to collapse the right tails for better readability. 

```{r warning=FALSE, message=FALSE, echo= FALSE, out.width="100%", fig.height=ridgeheight}
ggplot(df, aes(x = block_length, y = sample, fill = stat(x))) +
  geom_density_ridges_gradient(scale = 1.4) +
  scale_fill_viridis_c(option = "G", trans = "log10") +
  theme_minimal() +
  theme(legend.position = "none") +
  scale_x_log10(
     breaks = trans_breaks("log10", function(x) 10^x),
     labels = trans_format("log10", math_format(10^.x))
  ) +
  labs(title = "Distribution of Haplotype Lengths by Sample", fill = "Haplotype Length (bp)") +
  xlab("Haplotype Length (log scale)") +
  ylab("")
```
