---
title: "Leviathan Structural Variants Summary"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    df_print: kable
    toc: no
---

Below is a series of tables and plots detailing the structural variants
identified by `LEVIATHAN`. The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
```
```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(magrittr)
library(tidyr)
library(DT)
library(plotly)
library(circlize)
```

```{r echo = FALSE}
#fa.sizes <- read.table("~/genome.fasta.fai", header = F)[,1:2]
faidx <- snakemake@input[["faidx"]]
fa.sizes <- read.table(faidx, header = F)[,1:2] %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")
# limit the data to only the 30 largest contigs
if (nrow(fa.sizes) > 30){
  fa.sizes <- fa.sizes[1:30, ]
}

#sv <- do.call(rbind, lapply(c("~/1.sv.stats", "~/2.sv.stats"), function(x) read.table(x, header = T)))
sv <- do.call(rbind, lapply(snakemake@input[["statsfiles"]], function(x) read.table(x, header = T)))

if (nrow(sv) == 0) {
  print("No variants detected")
  q()
}
```

## Various information {.tabset}
### summary info
This table shows the overall counts, average per contig, and standard deviation per contig for detected structural variants across populations.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
genstats <- sv %>%
  group_by(population, contig, type) %>%
  summarise(tot = n()) %>%
  group_by(population, type) %>%
  summarise(count = sum(tot), avg = round(mean(tot, na.rm = T),2 ), sd = round(sd(tot, na.rm = T), 2)) %>% 
  pivot_wider(names_from = type, values_from = c(count, avg, sd), names_glue = "{type}.{.value}", names_vary = "slowest")

DT::datatable(genstats, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

### all variants
This table details all the variants LEVIATHAN detected and the quantity/quality of evidence for those detections.
```{r message = FALSE, echo = FALSE}
DT::datatable(sv, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

```{r echo = FALSE, warnings= FALSE, message = FALSE}
sv$length <- as.numeric(gsub("\\.", "1", sv$length))
```

## Per-contig information {.tabset}
### counts per contig
This table details counts each type of structural variant for every contig for every population.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grpcounts <- sv %>%
  group_by(population, contig, type) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = population, values_from = count)

DT::datatable(grpcounts, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

### bp per contig
This table details the lengths (in bp) of SV's for every contig for every population. Breakends are not counted because they do not have a size.
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grplens <- sv %>% 
  filter(type != "BND") %>%
  group_by(population, contig, type) %>%
  summarise(bp = sum(length)) %>%
  pivot_wider(names_from = population, values_from = bp)

DT::datatable(grplens, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

---

## SV size and location
The plots below are a visual representation of each SV where it was detected.
Contigs that are absent did not have any SV's detected on them. Within each 
contig, each population is plotted as a row to help visualize shared
variants between populations. Long population names create issues with row name
formatting, therefore the population name is included in the hover tooltip.
Plotting only the first 30 largest contigs.

```{r echo = FALSE, message = FALSE, warning = FALSE}
sv$ystart <- case_when(
  sv$type == "INV" ~ 0.1,
  sv$type == "DUP" ~ 1.1,
  sv$type == "DEL" ~ 2.1,
  sv$type == "BND" ~ 3.1,
)
sv$ystop <- sv$ystart + 0.8
```

```{r echo = FALSE, out.width = '100%', warning=FALSE}
color.palette <- c("DEL"="#004D40", "DUP"="#FFC107", "INV"="#1E88E5","BND"="#D81B60")
l <- htmltools::tagList()

for (i in 1:nrow(fa.sizes)) {
  sv.filt <- sv %>% filter(contig == fa.sizes$contig[i])
  sv_stats <- group_by(sv.filt, type) %>% summarise(n = length(type))
  if (nrow(sv_stats) == 0) {
    next
  }
  plt <- sv.filt %>%
    ggplot() +
    geom_rect(
      alpha = 0.7,
      aes(
        ymin = ystart,
        ymax = ystop,
        xmin = position_start,
        xmax = position_end,
        fill = type,
        color = type,
        text = sprintf("Population: %s<br>Type: %s<br>Position: %s-%s<br>barcodes: %s", population, type, position_start, position_end, n_barcodes)
        )
      ) +
    geom_hline(yintercept = 1:3, color = "grey90") +
    scale_color_manual(values = color.palette) +
    scale_fill_manual(values = color.palette) +
    facet_grid(rows = vars(population)) +
    theme_light() +
    theme(
      axis.text.y = element_text(angle=90),
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_blank(),
      legend.position = "none",
    ) +
    xlim(0, fa.sizes$size[i]) +
    scale_y_continuous(
      breaks = c(0.5, 1.5, 2.5, 3.5),
      labels = c("INV", "DUP", "DEL", "BND"),
      limits = c(0,4)
    ) +
    coord_cartesian(xlim = c(0, fa.sizes$size[i] + 1), expand = F) +
    xlab("Position (bp)") +
    labs(fill = "SV type", color = "SV type") +
    ggtitle(fa.sizes$contig[i])
  l[[i]] <- ggplotly(plt, tooltip = "text")
}
l
```