<!DOCTYPE html>
<html lang="en" class="h-full">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="turbo-root" content="/HARPY">
    <meta name="turbo-cache-control" content="no-cache" data-turbo-track="reload" data-track-token="3.5.0.765493029486">

    <!-- See retype.com -->
    <meta name="generator" content="Retype 3.5.0">

    <!-- Primary Meta Tags -->
    <title> Simulate Genomic Variants | Harpy</title>
    <meta name="title" content=" Simulate Genomic Variants | Harpy">
    <meta name="description" content="Simulate snps, indels, inversions, cnv, translocations">

    <!-- Canonical -->
    <link rel="canonical" href="https://pdimens.github.io/HARPY/modules/simulate/simulate-variants/">

    <!-- Open Graph / Facebook -->
    <meta property="og:type" content="website">
    <meta property="og:url" content="https://pdimens.github.io/HARPY/modules/simulate/simulate-variants/">
    <meta property="og:title" content=" Simulate Genomic Variants | Harpy">
    <meta property="og:description" content="Simulate snps, indels, inversions, cnv, translocations">

    <!-- Twitter -->
    <meta property="twitter:card" content="summary_large_image">
    <meta property="twitter:url" content="https://pdimens.github.io/HARPY/modules/simulate/simulate-variants/">
    <meta property="twitter:title" content=" Simulate Genomic Variants | Harpy">
    <meta property="twitter:description" content="Simulate snps, indels, inversions, cnv, translocations">

    <script data-cfasync="false">(function () { var el = document.documentElement, m = localStorage.getItem("doc_theme"), wm = window.matchMedia; if (m === "dark" || (!m && wm && wm("(prefers-color-scheme: dark)").matches)) { el.classList.add("dark") } else { el.classList.remove("dark") } })();</script>

    <link href="../../../static/favicon.png" rel="icon">
    <link href="../../../resources/css/retype.css?v=3.5.0.765493029486" rel="stylesheet">

    <script data-cfasync="false" src="../../../resources/js/config.js?v=3.5.0.765493029486" data-turbo-eval="false" defer></script>
    <script data-cfasync="false" src="../../../resources/js/retype.js?v=3.5.0" data-turbo-eval="false" defer></script>
    <script id="lunr-js" data-cfasync="false" src="../../../resources/js/lunr.js?v=3.5.0.765493029486" data-turbo-eval="false" defer></script>
    <script id="prism-js" data-cfasync="false" src="../../../resources/js/prism.js?v=3.5.0.765493029486" defer></script>
    <script id="mermaid-js" data-cfasync="false" src="../../../resources/js/mermaid.js?v=3.5.0" defer></script>
</head>
<body>
    <div id="docs-app" class="relative text-base antialiased text-gray-700 bg-white font-body dark:bg-dark-850 dark:text-dark-300">
        <div class="absolute bottom-0 left-0 bg-gray-100 dark:bg-dark-800" style="top: 5rem; right: 50%"></div>
    
        <header id="docs-site-header" class="sticky top-0 z-30 flex w-full h-16 bg-white border-b border-gray-200 md:h-20 dark:bg-dark-850 dark:border-dark-650">
            <div class="container relative flex items-center justify-between pr-6 grow md:justify-start">
                <!-- Mobile menu button skeleton -->
                <button v-cloak class="skeleton docs-mobile-menu-button flex items-center justify-center shrink-0 overflow-hidden dark:text-white focus:outline-none rounded-full w-10 h-10 ml-3.5 md:hidden"><svg xmlns="http://www.w3.org/2000/svg" class="mb-px shrink-0" width="24" height="24" viewBox="0 0 24 24" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor"><path d="M2 4h20v2H2zM2 11h20v2H2zM2 18h20v2H2z"></path></g></svg></button>
                <div v-cloak id="docs-sidebar-toggle"></div>
        
                <!-- Logo -->
                <div class="flex items-center justify-between h-full py-2 md:w-75">
                    <div class="flex items-center px-2 md:px-6">
                        <a id="docs-site-logo" href="../../../" class="flex items-center leading-snug text-2xl">
                            <span class="w-10 mr-2 grow-0 shrink-0 overflow-hidden">
                                <img class="max-h-10 dark:hidden md:inline-block" src="../../../static/favicon.png">
                                <img class="max-h-10 hidden dark:inline-block" src="../../../static/favicon.png">
                            </span>
                            <span class="dark:text-white font-semibold line-clamp-1 md:line-clamp-2">Harpy</span>
                        </a><span class="hidden px-2 py-1 ml-4 text-sm font-semibold leading-none text-root-logo-label-text bg-root-logo-label-bg rounded-sm md:inline-block">v0.9.x</span>
                    </div>
        
                    <span class="hidden h-8 border-r md:inline-block dark:border-dark-650"></span>
                </div>
        
                <div class="flex justify-between md:grow">
                    <!-- Top Nav -->
                    <nav class="hidden md:flex">
                        <ul class="flex flex-col mb-4 md:pl-16 md:mb-0 md:flex-row md:items-center">
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/pdimens/HARPY">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <symbol id="mark-github-icon-symbol" viewBox="0 0 16 16"><path d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"/></symbol><use xlink:href="#mark-github-icon-symbol" />
                                        </g>
                                    </svg>
                                    <span>GitHub</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.therkildsenlab.com/">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M8 8.807V3.5h-.563a.75.75 0 0 1 0-1.5h9.125a.75.75 0 0 1 0 1.5H16v5.307l5.125 9.301c.964 1.75-.302 3.892-2.299 3.892H5.174c-1.998 0-3.263-2.142-2.3-3.892ZM4.189 18.832a1.123 1.123 0 0 0 .985 1.668h13.652a1.123 1.123 0 0 0 .985-1.668L17.7 15H6.3ZM14.5 3.5h-5V9a.75.75 0 0 1-.093.362L7.127 13.5h9.746l-2.28-4.138A.75.75 0 0 1 14.5 9Z"/>
                                        </g>
                                    </svg>
                                    <span>Therkildsen Lab</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.fml.tuebingen.mpg.de/9418/haplotagging">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M5.873 3.26A.748.748 0 0 1 6.44 3h11.31c.223 0 .434.099.576.27l5 6a.75.75 0 0 1-.028.992l-10.75 11.5a.75.75 0 0 1-1.096 0l-10.75-11.5a.75.75 0 0 1-.02-1.003l5.19-6Zm.91 1.24L2.258 9.73 12 20.153l9.75-10.43L17.399 4.5Z"/>
                                        </g>
                                    </svg>
                                    <span>Haplotag Home</span>
                                </a>
                            </li>
        
                        </ul>
                    </nav>
        
                    <!-- Header Right Skeleton -->
                    <div v-cloak class="flex justify-end grow skeleton">
        
                        <!-- Search input mock -->
                        <div class="relative hidden w-40 lg:block lg:max-w-sm lg:ml-auto">
                            <div class="absolute flex items-center justify-center h-full pl-3 dark:text-dark-300">
                                <svg xmlns="http://www.w3.org/2000/svg" class="icon-base" width="16" height="16" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 1px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                            </div>
                            <input class="w-full h-10 placeholder-gray-400 transition-colors duration-200 ease-in bg-gray-200 border border-transparent rounded md:text-sm hover:bg-white hover:border-gray-300 focus:outline-none focus:bg-white focus:border-gray-500 dark:bg-dark-600 dark:border-dark-600 dark:placeholder-dark-400" style="padding: 0.625rem 0.75rem 0.625rem 2rem" type="text" placeholder="Search">
                        </div>
        
                        <!-- Mobile search button -->
                        <div class="flex items-center justify-center w-10 h-10 lg:hidden">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="20" height="20" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                        </div>
        
                        <!-- Dark mode switch placeholder -->
                        <div class="w-10 h-10 lg:ml-2"></div>
        
                        <!-- History button -->
                        <div class="flex items-center justify-center w-10 h-10" style="margin-right: -0.625rem;">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="22" height="22" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><g ><path d="M12.01 6.01c-.55 0-1 .45-1 1V12a1 1 0 00.4.8l3 2.22a.985.985 0 001.39-.2.996.996 0 00-.21-1.4l-2.6-1.92V7.01c.02-.55-.43-1-.98-1z"></path><path d="M12.01 1.91c-5.33 0-9.69 4.16-10.05 9.4l-.29-.26a.997.997 0 10-1.34 1.48l1.97 1.79c.19.17.43.26.67.26s.48-.09.67-.26l1.97-1.79a.997.997 0 10-1.34-1.48l-.31.28c.34-4.14 3.82-7.41 8.05-7.41 4.46 0 8.08 3.63 8.08 8.09s-3.63 8.08-8.08 8.08c-2.18 0-4.22-.85-5.75-2.4a.996.996 0 10-1.42 1.4 10.02 10.02 0 007.17 2.99c5.56 0 10.08-4.52 10.08-10.08.01-5.56-4.52-10.09-10.08-10.09z"></path></g></g></svg>
                        </div>
                    </div>
        
                    <div v-cloak class="flex justify-end grow">
                        <div id="docs-mobile-search-button"></div>
                        <doc-search-desktop></doc-search-desktop>
        
                        <doc-theme-switch class="lg:ml-2"></doc-theme-switch>
                        <doc-history></doc-history>
                    </div>
                </div>
            </div>
        </header>
    
        <div class="container relative flex bg-white">
            <!-- Sidebar Skeleton -->
            <div v-cloak class="fixed flex flex-col shrink-0 duration-300 ease-in-out bg-gray-100 border-gray-200 sidebar top-20 w-75 border-r h-screen md:sticky transition-transform skeleton dark:bg-dark-800 dark:border-dark-650">
            
                <!-- Render this div, if config.showSidebarFilter is `true` -->
                <div class="flex items-center h-16 px-6">
                    <input class="w-full h-8 px-3 py-2 transition-colors duration-200 ease-linear bg-white border border-gray-200 rounded shadow-none text-sm focus:outline-none focus:border-gray-600 dark:bg-dark-600 dark:border-dark-600" type="text" placeholder="Filter">
                </div>
            
                <div class="pl-6 mt-1 mb-4">
                    <div class="w-32 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-48 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-40 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-32 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-48 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-40 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                </div>
            
                <div class="shrink-0 mt-auto bg-transparent dark:border-dark-650">
                    <a class="flex items-center justify-center flex-nowrap h-16 text-gray-400 dark:text-dark-400 hover:text-gray-700 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                        <span class="text-xs whitespace-nowrap">Powered by</span>
                        <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                    </a>
                </div>
            </div>
            
            <!-- Sidebar component -->
            <doc-sidebar v-cloak>
                <template #sidebar-footer>
                    <div class="shrink-0 mt-auto border-t md:bg-transparent md:border-none dark:border-dark-650">
            
                        <div class="py-3 px-6 md:hidden border-b dark:border-dark-650">
                            <nav>
                                <ul class="flex flex-wrap justify-center items-center">
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/pdimens/HARPY">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <symbol id="mark-github-icon-symbol" viewBox="0 0 16 16"><path d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"/></symbol><use xlink:href="#mark-github-icon-symbol" />
                                                </g>
                                            </svg>
                                            <span>GitHub</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.therkildsenlab.com/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M8 8.807V3.5h-.563a.75.75 0 0 1 0-1.5h9.125a.75.75 0 0 1 0 1.5H16v5.307l5.125 9.301c.964 1.75-.302 3.892-2.299 3.892H5.174c-1.998 0-3.263-2.142-2.3-3.892ZM4.189 18.832a1.123 1.123 0 0 0 .985 1.668h13.652a1.123 1.123 0 0 0 .985-1.668L17.7 15H6.3ZM14.5 3.5h-5V9a.75.75 0 0 1-.093.362L7.127 13.5h9.746l-2.28-4.138A.75.75 0 0 1 14.5 9Z"/>
                                                </g>
                                            </svg>
                                            <span>Therkildsen Lab</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.fml.tuebingen.mpg.de/9418/haplotagging">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M5.873 3.26A.748.748 0 0 1 6.44 3h11.31c.223 0 .434.099.576.27l5 6a.75.75 0 0 1-.028.992l-10.75 11.5a.75.75 0 0 1-1.096 0l-10.75-11.5a.75.75 0 0 1-.02-1.003l5.19-6Zm.91 1.24L2.258 9.73 12 20.153l9.75-10.43L17.399 4.5Z"/>
                                                </g>
                                            </svg>
                                            <span>Haplotag Home</span>
                                        </a>
                                    </li>
            
                                </ul>
                            </nav>
                        </div>
            
                        <a class="flex items-center justify-center flex-nowrap h-16 text-gray-400 dark:text-dark-400 hover:text-gray-700 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                            <span class="text-xs whitespace-nowrap">Powered by</span>
                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                        </a>
                    </div>
                </template>
            </doc-sidebar>
    
            <div class="grow min-w-0 dark:bg-dark-850">
                <!-- Render "toolbar" template here on api pages --><!-- Render page content -->
                <div class="flex">
                    <div class="min-w-0 p-4 grow md:px-16">
                        <main class="relative pb-12 lg:pt-2">
                            <div class="docs-markdown" id="docs-content">
                                <!-- Rendered if sidebar right is enabled -->
                                <div id="docs-sidebar-right-toggle"></div>
                                <!-- Page content  -->
<doc-anchor-target id="simulate-genomic-variants" class="break-words">
    <h1>
        <doc-anchor-trigger class="header-anchor-trigger" to="#simulate-genomic-variants">#</doc-anchor-trigger>
        <span><svg xmlns="http://www.w3.org/2000/svg" class="docs-icon" viewBox="0 0 24 24" role="presentation"><g fill="currentColor"><g><path d="M14.265 1.627c0 3.545 1.869 5.327 3.479 7.021 1.54 1.62 3.006 3.163 3.006 6.102 0 4.812-3.753 8.25-8.565 8.25-4.813 0-8.935-3.421-8.935-8.25 0-2.039.962-4.011 2.509-4.899.305-.175.672.007.803.334C7.563 12.684 8.797 12.64 9.437 12c.388-.387.47-1.116-.004-2.062-2.405-4.812 1.863-8.279 4.2-8.854.336-.082.615.198.632.543ZM12.185 21.5c4.059 0 7.065-2.84 7.065-6.75 0-2.337-1.093-3.489-2.678-5.158l-.021-.023c-1.44-1.517-3.139-3.351-3.649-6.557a6.148 6.148 0 0 0-1.911 1.76c-.787 1.144-1.147 2.633-.216 4.495.603 1.205.777 2.74-.277 3.794-.657.657-1.762 1.1-2.956.586-.752-.324-1.353-.955-1.838-1.79-.567.706-.954 1.74-.954 2.893 0 3.847 3.288 6.75 7.435 6.75Z"/></g></g></svg> Simulate Genomic Variants</span>
    </h1>
</doc-anchor-target>
<p>Simulate snps, indels, inversions, cnv, translocations</p>
<doc-panel id="you-will-need">
    <template #title><svg xmlns="http://www.w3.org/2000/svg" class="docs-icon" viewBox="0 0 24 24" role="presentation"><g fill="currentColor"><g><path d="M3.5 3.75a.25.25 0 0 1 .25-.25h13.5a.25.25 0 0 1 .25.25v10a.75.75 0 0 0 1.5 0v-10A1.75 1.75 0 0 0 17.25 2H3.75A1.75 1.75 0 0 0 2 3.75v16.5c0 .966.784 1.75 1.75 1.75h7a.75.75 0 0 0 0-1.5h-7a.25.25 0 0 1-.25-.25V3.75Z"/><path d="M6.25 7a.75.75 0 0 0 0 1.5h8.5a.75.75 0 0 0 0-1.5h-8.5Zm-.75 4.75a.75.75 0 0 1 .75-.75h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1-.75-.75Zm16.28 4.53a.75.75 0 1 0-1.06-1.06l-4.97 4.97-1.97-1.97a.75.75 0 1 0-1.06 1.06l2.5 2.5a.75.75 0 0 0 1.06 0l5.5-5.5Z"/></g></g></svg> You will need</template>
<ul>
<li>a reference genome in fasta or gzipped fasta format</li>
</ul>
</doc-panel>
<p>You may want to benchmark haplotag data on different kinds of genomic variants. To
do that, you&#x27;ll need <em>known</em> variants, and typically simulations are how you achieve
that. This series of modules simulates genomic variants onto a genome, either randomly
or specific variants provided in VCF files. The simulator Harpy uses,
<a href="https://github.com/yjx1217/simuG">simuG</a>, can only simulate one type of
variant at a time and each variant type has their own set of parameters. This page
is divided by variant types to help you navigate the process. The general usage
for simulating variants is:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<div class="codeblock-title">usage</div>
<pre class="language-bash"><code v-pre class="language-bash">harpy simulate variant OPTIONS... INPUT_GENOME</code></pre>
</doc-codeblock></div>
<div class="codeblock-wrapper"><doc-codeblock>
<div class="codeblock-title">example</div>
<pre class="language-bash"><code v-pre class="language-bash">harpy simulate inversion -n 10 --min-size 1000 --max-size 50000  path/to/genome.fasta</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="modules">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#modules">#</doc-anchor-trigger>
        <span>Modules</span>
    </h2>
</doc-anchor-target>
<p>There are 4 submodules with very obvious names:</p>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="comfortable">
<thead>
<tr>
<th style="text-align: left;">submodule</th>
<th style="text-align: left;">what it does</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code v-pre>snpindel</code></td>
<td style="text-align: left;">simulates single nucleotide polymorphisms (snps) and insertion-deletions (indels)</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>inversion</code></td>
<td style="text-align: left;">simulates inversions</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>cnv</code></td>
<td style="text-align: left;">simulates copy number variants</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>translocation</code></td>
<td style="text-align: left;">simulates translocations</td>
</tr>
</tbody>
</table>
</div>
<doc-anchor-target id="running-options">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#running-options">#</doc-anchor-trigger>
        <span><svg xmlns="http://www.w3.org/2000/svg" class="docs-icon" viewBox="0 0 24 24" role="presentation"><g fill="currentColor"><g><path d="M9.25 12a.75.75 0 0 1-.22.53l-2.75 2.75a.75.75 0 0 1-1.06-1.06L7.44 12 5.22 9.78a.75.75 0 1 1 1.06-1.06l2.75 2.75c.141.14.22.331.22.53Zm2 2a.75.75 0 0 0 0 1.5h5a.75.75 0 0 0 0-1.5h-5Z"/><path d="M0 4.75C0 3.784.784 3 1.75 3h20.5c.966 0 1.75.784 1.75 1.75v14.5A1.75 1.75 0 0 1 22.25 21H1.75A1.75 1.75 0 0 1 0 19.25Zm1.75-.25a.25.25 0 0 0-.25.25v14.5c0 .138.112.25.25.25h20.5a.25.25 0 0 0 .25-.25V4.75a.25.25 0 0 0-.25-.25Z"/></g></g></svg> Running Options</span>
    </h2>
</doc-anchor-target>
<p>While there are serveral differences between the submodule command line options, each has available all the
<a href="../../../commonoptions/">common runtime options</a> like other Harpy modules. Each requires and input genome at the
end of the command line, and each requires either a <code v-pre>--count</code> of variants to randomly simulate, or a <code v-pre>--vcf</code> of
specific variants to simulate. There are also these unifying options among the different variant types:</p>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="comfortable">
<thead>
<tr>
<th style="text-align: left;">argument</th>
<th style="text-align: left;">short name</th>
<th style="text-align: left;">type</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code v-pre>INPUT_GENOME</code></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">file path</td>
<td style="text-align: left;">The haploid genome to simulate variants onto. <strong>REQUIRED</strong></td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--prefix</code></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">string</td>
<td style="text-align: left;">Naming prefix for output files (default: <code v-pre>sim.{module_name}</code>)</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--exclude-chr</code></td>
<td style="text-align: left;"><code v-pre>-e</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: left;">Text file of chromosomes to avoid, one per line</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--centromeres</code></td>
<td style="text-align: left;"><code v-pre>-c</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: left;">GFF3 file of centromeres to avoid</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--genes</code></td>
<td style="text-align: left;"><code v-pre>-g</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: left;">GFF3 file of genes to avoid simulating over (see <code v-pre>snpindel</code> for caveat)</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--heterozygosity</code></td>
<td style="text-align: left;"><code v-pre>-z</code></td>
<td style="text-align: left;">float between [0,1]</td>
<td style="text-align: left;"><doc-anchor-trigger to="#heterozygosity">% heterozygosity to simulate diploid later</doc-anchor-trigger> (default: <code v-pre>0</code>)</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--randomseed</code></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">Random seed for simulation</td>
</tr>
</tbody>
</table>
</div>
<doc-panels>
   <doc-panel id="snps-and-indels" collapsed>
       <template #title>snps and indels</template>
<doc-anchor-target id="snpindel">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#snpindel">#</doc-anchor-trigger>
        <span>snpindel</span>
    </h3>
</doc-anchor-target>
<div class="flex mb-6">
    <div class="shrink-0 w-1 rounded-tl rounded-bl bg-yellow-500"></div>
    <div class="flex w-full py-4 border border-l-0 border-gray-300 rounded-tr rounded-br doc-alert bg-white dark:bg-dark-700 dark:border-dark-700" role="alert">
        <div class="flex items-center ml-4 h-7">
            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px text-yellow-500" width="22" height="22" viewBox="0 0 24 24" role="presentation">
                <g fill="currentColor"><g>
                    <path d="M22.48 15.59L14.01 1.45A2.968 2.968 0 0012.16.09c-.78-.19-1.58-.07-2.27.35-.41.25-.76.6-1.01 1.01v.01L.4 15.6c-.83 1.43-.33 3.27 1.1 4.1.45.26.95.4 1.48.4h16.95c.8-.01 1.55-.33 2.11-.9.56-.57.87-1.33.86-2.13a3.04 3.04 0 00-.42-1.48zm-1.87 2.21c-.19.19-.44.3-.69.3H2.99c-.17 0-.34-.05-.49-.13a.992.992 0 01-.37-1.35L10.6 2.48c.08-.14.2-.25.34-.33a.992.992 0 011.37.33l8.46 14.13c.09.15.13.32.13.49 0 .26-.1.51-.29.7z"></path>
                    <path d="M11.45 12.1c.55 0 1-.45 1-1v-4c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .56.45 1 1 1zM11.46 14.1c-.56 0-1 .45-1 1s.45 1 1 1 1-.45 1-1-.45-1-1-1z"></path>
                </g></g>
            </svg>
        </div>
        <div class="pr-5 ml-3 w-full">
            <h5>SNPs can be slow</h5>
<p>Given software limitations, simulating many SNPs (&gt;10,000) will be noticeably slower than the other variant types.</p>
        </div>
    </div>
</div>
<p>A single nucleotide polymorphism (&quot;SNP&quot;) is a genomic variant at a single base position in the DNA (<a href="https://www.genome.gov/genetics-glossary/Single-Nucleotide-Polymorphisms">source</a>).
An indel, is a type of mutation that involves the addition/deletion of one or more nucleotides into a segment of DNA (<a href="https://www.genome.gov/genetics-glossary/Insertion">insertions</a>, <a href="https://www.genome.gov/genetics-glossary/Deletion">deletions</a>).
The snp and indel variants are combined in this module because <code v-pre>simuG</code> allows simulating them together. The
ratio parameters control different things for snp and indel variants and have special meanings when setting
the value to either <code v-pre>9999</code> or <code v-pre>0</code> :</p>
<ul>
<li><code v-pre>--titv-ratio</code>
<ul>
<li><code v-pre>9999</code>: transitions only</li>
<li><code v-pre>0</code>: transversions only</li>
</ul>
</li>
<li><code v-pre>--indel-ratio</code>
<ul>
<li><code v-pre>9999</code>: insertions only</li>
<li><code v-pre>0</code>: deletions only</li>
</ul>
</li>
</ul>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="comfortable">
<thead>
<tr>
<th style="text-align: left;">argument</th>
<th style="text-align: center;">short name</th>
<th style="text-align: left;">type</th>
<th style="text-align: center;">default</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code v-pre>--snp-vcf</code></td>
<td style="text-align: center;"><code v-pre>-v</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: center;"></td>
<td style="text-align: left;">VCF file of known snps to simulate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--indel-vcf</code></td>
<td style="text-align: center;"><code v-pre>-i</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: center;"></td>
<td style="text-align: left;">VCF file of known indels to simulate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--snp-count</code></td>
<td style="text-align: center;"><code v-pre>-n</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">0</td>
<td style="text-align: left;">Number of random snps to simluate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--indel-count</code></td>
<td style="text-align: center;"><code v-pre>-m</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">0</td>
<td style="text-align: left;">Number of random indels to simluate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--titv-ratio</code></td>
<td style="text-align: center;"><code v-pre>-r</code></td>
<td style="text-align: left;">float</td>
<td style="text-align: center;">0.5</td>
<td style="text-align: left;">Transition/Transversion ratio for snps</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--indel-ratio</code></td>
<td style="text-align: center;"><code v-pre>-d</code></td>
<td style="text-align: left;">float</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Insertion/Deletion ratio for indels</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--indel-size-alpha</code></td>
<td style="text-align: center;"><code v-pre>-a</code></td>
<td style="text-align: left;">float</td>
<td style="text-align: center;">2.0</td>
<td style="text-align: left;">Exponent Alpha for power-law-fitted indel size distribution</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--indel-size-constant</code></td>
<td style="text-align: center;"><code v-pre>-l</code></td>
<td style="text-align: left;">float</td>
<td style="text-align: center;">0.5</td>
<td style="text-align: left;">Exponent constant for power-law-fitted indel size distribution</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--snp-gene-constraints</code></td>
<td style="text-align: center;"><code v-pre>-y</code></td>
<td style="text-align: left;">string</td>
<td style="text-align: center;"></td>
<td style="text-align: left;">How to constrain randomly simulated SNPs {<code v-pre>noncoding</code>,<code v-pre>coding</code>,<code v-pre>2d</code>,<code v-pre>4d</code>} when using <code v-pre>--genes</code></td>
</tr>
</tbody>
</table>
</div>
   </doc-panel>
   <doc-panel id="inversions" collapsed>
       <template #title>inversions</template>
<doc-anchor-target id="inversion">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#inversion">#</doc-anchor-trigger>
        <span>inversion</span>
    </h3>
</doc-anchor-target>
<p>Inversions are when a section of a chromosome appears in the reverse orientation (<a href="https://www.genome.gov/genetics-glossary/Inversion">source</a>).</p>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="comfortable">
<thead>
<tr>
<th style="text-align: left;">argument</th>
<th style="text-align: center;">short name</th>
<th style="text-align: left;">type</th>
<th style="text-align: center;">default</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code v-pre>--vcf</code></td>
<td style="text-align: center;"><code v-pre>-v</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: center;"></td>
<td style="text-align: left;">VCF file of known inversions to simulate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--count</code></td>
<td style="text-align: center;"><code v-pre>-n</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">0</td>
<td style="text-align: left;">Number of random inversions to simluate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--min-size</code></td>
<td style="text-align: center;"><code v-pre>-m</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">1000</td>
<td style="text-align: left;">Minimum inversion size (bp)</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--max-size</code></td>
<td style="text-align: center;"><code v-pre>-x</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">100000</td>
<td style="text-align: left;">Maximum inversion size (bp)</td>
</tr>
</tbody>
</table>
</div>
   </doc-panel>
   <doc-panel id="copy-number-variants" collapsed>
       <template #title>copy number variants</template>
<doc-anchor-target id="cnv">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#cnv">#</doc-anchor-trigger>
        <span>cnv</span>
    </h3>
</doc-anchor-target>
<p>A copy number variation (CNV) is when the number of copies of a particular gene varies
between individuals (<a href="https://www.genome.gov/genetics-glossary/Copy-Number-Variation">source</a>)
The ratio parameters control different things and have special meanings when setting
the value to either <code v-pre>9999</code> or <code v-pre>0</code> :</p>
<ul>
<li><code v-pre>--dup-ratio</code>
<ul>
<li><code v-pre>9999</code>: tandem duplications only</li>
<li><code v-pre>0</code>: dispersed duplications only</li>
</ul>
</li>
<li><code v-pre>--gain-ratio</code>
<ul>
<li><code v-pre>9999</code>: gain only</li>
<li><code v-pre>0</code>: loss only</li>
</ul>
</li>
</ul>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="comfortable">
<thead>
<tr>
<th style="text-align: left;">argument</th>
<th style="text-align: center;">short name</th>
<th style="text-align: left;">type</th>
<th style="text-align: center;">default</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code v-pre>--vcf</code></td>
<td style="text-align: center;"><code v-pre>-v</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: center;"></td>
<td style="text-align: left;">VCF file of known copy number variants to simulate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--count</code></td>
<td style="text-align: center;"><code v-pre>-n</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">0</td>
<td style="text-align: left;">Number of random cnv to simluate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--min-size</code></td>
<td style="text-align: center;"><code v-pre>-m</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">1000</td>
<td style="text-align: left;">Minimum cnv size (bp)</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--max-size</code></td>
<td style="text-align: center;"><code v-pre>-x</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">100000</td>
<td style="text-align: left;">Maximum cnv size (bp)</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--max-copy</code></td>
<td style="text-align: center;"><code v-pre>-y</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">10</td>
<td style="text-align: left;">Maximum number of copies</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--dup-ratio</code></td>
<td style="text-align: center;"><code v-pre>-d</code></td>
<td style="text-align: left;">float</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Tandem/Dispersed duplication ratio</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--gain-ratio</code></td>
<td style="text-align: center;"><code v-pre>-l</code></td>
<td style="text-align: left;">float</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Relative ratio of DNA gain over DNA loss</td>
</tr>
</tbody>
</table>
</div>
   </doc-panel>
   <doc-panel id="translocations" collapsed>
       <template #title>translocations</template>
<doc-anchor-target id="translocation">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#translocation">#</doc-anchor-trigger>
        <span>translocation</span>
    </h3>
</doc-anchor-target>
<p>A translocation occurs when a chromosome breaks and the fragmented pieces re-attach to different chromosomes (<a href="https://www.genome.gov/genetics-glossary/Translocation">source</a>).</p>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="comfortable">
<thead>
<tr>
<th style="text-align: left;">argument</th>
<th style="text-align: center;">short name</th>
<th style="text-align: left;">type</th>
<th style="text-align: center;">default</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code v-pre>--vcf</code></td>
<td style="text-align: center;"><code v-pre>-v</code></td>
<td style="text-align: left;">file path</td>
<td style="text-align: center;"></td>
<td style="text-align: left;">VCF file of known inversions to simulate</td>
</tr>
<tr>
<td style="text-align: left;"><code v-pre>--count</code></td>
<td style="text-align: center;"><code v-pre>-n</code></td>
<td style="text-align: left;">integer</td>
<td style="text-align: center;">0</td>
<td style="text-align: left;">Number of random inversions to simluate</td>
</tr>
</tbody>
</table>
</div>
   </doc-panel>
</doc-panels>
<doc-anchor-target id="simulate-known-variants">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#simulate-known-variants">#</doc-anchor-trigger>
        <span>Simulate known variants</span>
    </h2>
</doc-anchor-target>
<p>Rather than simulating random variants, you can use a VCF file as input to any of the submodules
to have <code v-pre>simuG</code> simulate the variants (of that type) from the VCF file. This becomes particularly
handy because the modules output a VCF file of the variants that were introduced, which you can
modify and reuse as you see fit (see <doc-anchor-trigger to="#heterozygosity">heterozygosity</doc-anchor-trigger>). Using <code v-pre>--genes</code>,
<code v-pre>--centromeres</code>, or <code v-pre>--exclude-chr</code> would still avoid creating variants in those regions as with
random simulation, except with SNPs, where you would have to specify the contraints for using
<code v-pre>--genes</code> as per usual.</p>
<doc-anchor-target id="heterozygosity">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#heterozygosity">#</doc-anchor-trigger>
        <span>Heterozygosity</span>
    </h2>
</doc-anchor-target>
<p>Each submodule has a <code v-pre>--heterozygosity</code> parameter where you can specify the heterozygosity of
an intended diploid genome, should you use the resulting VCF(s) to simulate variants again.
This <strong>does not</strong> create a diploid genome for you, rather, it
takes the output VCF from <code v-pre>simuG</code> and creates two new VCF files (<code v-pre>{prefix}.hap1.vcf</code>,
<code v-pre>{prefix}.hap2.vcf</code>) that have their variants shuffled between the two haplotypes to
achieve the desired heterozygosity. You can then use the two resulting VCF files to
<doc-anchor-trigger to="#simulate-known-variants">simulate known variants</doc-anchor-trigger> onto a genome and create a diploid genome!
A simplified workflow is provided below to explain that in better context.</p>
<doc-panel id="how-the-paramater-is-actually-used" collapsed>
    <template #title>How the paramater is actually used</template>
<p>To understand how heterozygosity is created from the <code v-pre>simuG</code> VCF output, consider a genome
with 5 variants added to it, here represented as a column labelled <code v-pre>h1</code> with <code v-pre>1</code> being the presence
of a variant (the ALT allele).</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">h1
1
1
1
1
1</code></pre>
</doc-codeblock></div>
<p>If we were to simulate those same variants onto the genome again, it would create a homozygote
at every position (<code v-pre>h2</code> is the second haplotype):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">h1 h2
1  1
1  1
1  1
1  1
1  1</code></pre>
</doc-codeblock></div>
<p>However, if we omit some of the variants on <code v-pre>h2</code> to create 40% heterozygosity (2/5),
we would now have heterozygotes, except the ALT allele for the heterozygote would
only every be on the first haplotype <code v-pre>h1</code>:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">h1 h2
1  1
1  1
1  1
1     &lt;- heterozygote with ALT on h1
1     &lt;- heterozygote with ALT on h1</code></pre>
</doc-codeblock></div>
<p>It would probably be more biologically sound to then make sure that the ALT allele
in the heterozygote can appear in either haplotype:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">h1 h2
1  1
1  1
1  1
   1  &lt;- heterozygote with ALT on h2
1     &lt;- heterozygote with ALT on h1</code></pre>
</doc-codeblock></div>
<p>Within Harpy, a heterozygous variant has a 50% chance of being assigned to one of the haplotypes.
So that&#x27;s the logic behind the <code v-pre>--heterozygosity</code> parameter and why it ouputs 3 VCF files:</p>
<ol>
<li>the VCF <code v-pre>simuG</code> outputs of variants added to the genome</li>
<li>haplotype 1 of that VCF file with some of the variants</li>
<li>haplotype 2 of that VCF file with some of the variants</li>
</ol>
<p>Knowing that, you can then have a workflow to start with a haploid assembly and create
a diploid assembly with simulated variants.</p>
</doc-panel>
<doc-anchor-target id="simulate-diploid-assembly">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#simulate-diploid-assembly">#</doc-anchor-trigger>
        <span>Simulate Diploid Assembly</span>
    </h2>
</doc-anchor-target>
<p>Here is a simple but realistic workflow of creating a diploid assembly with simulated variants.
If you haven&#x27;t already, please read the sections about <doc-anchor-trigger to="#simulate-known-variants">simulating known variants</doc-anchor-trigger>
and <doc-anchor-trigger to="#heterozygosity">heterozygosity</doc-anchor-trigger>. The idea here is that due to the limitations of <code v-pre>simuG</code>, we can
only simulate one type of variant at a time and we will take advantage of the VCF files created by
the <code v-pre>--heterozygosity</code> parameter to simulate known variants from random ones! The basic idea is such:</p>
<doc-anchor-target id="step-1">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#step-1">#</doc-anchor-trigger>
        <span>Step 1</span>
    </h4>
</doc-anchor-target>
<p>Simulate random variants onto your haploid assembly with <code v-pre>--heterozygosity</code> (<code v-pre>-z</code>) set above <code v-pre>0</code>.
We aren&#x27;t interested in the resulting genome, but rather the positions of the variants <code v-pre>simuG</code> created.</p>
<div class="overflow-hidden mb-6 py-4 dark:px-5 dark:bg-white rounded-md"><pre class="mermaid w-full">graph LR
    geno(haploid genome)--&gt;|simulate inversion -n 10 -z 0.5|hap(inversions.vcf)
    hap--&gt;hap1(inversion.hap1.vcf)
    hap--&gt;hap2(inversion.hap2.vcf)</pre></div>
<doc-anchor-target id="step-2">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#step-2">#</doc-anchor-trigger>
        <span>Step 2</span>
    </h4>
</doc-anchor-target>
<p>Use the resulting hap1 and hap2 VCF files to simulate those same variants, but shuffled
into homozygotes and heterozygotes, onto the original haploid genome, creating two haplotype
genomes.</p>
<div class="overflow-hidden mb-6 py-4 dark:px-5 dark:bg-white rounded-md"><pre class="mermaid w-full">graph LR
    subgraph id1 [&quot;Inputs&quot;]
    hap1(inversion.hap1.vcf)---geno(haploid genome)
    end
    id1--&gt;|simulate inversion -v|hapgeno(haplotype-1 genome)
    style id1 fill:#f0f0f0,stroke:#e8e8e8,stroke-width:2px</pre></div>
<doc-anchor-target id="step-3">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#step-3">#</doc-anchor-trigger>
        <span>Step 3</span>
    </h4>
</doc-anchor-target>
<p>Use the one of the new genome haplotypes for simulating other kinds of variants.
Again, use <code v-pre>--heterozygosity</code> (<code v-pre>-z</code>) with a value greater than <code v-pre>0</code>. Like <doc-anchor-trigger to="#step-1"><strong>Step 1</strong></doc-anchor-trigger>,
we&#x27;re only interested in the haplotype VCF files (positions of variants) and not the resulting
genome.</p>
<div class="overflow-hidden mb-6 py-4 dark:px-5 dark:bg-white rounded-md"><pre class="mermaid w-full">graph LR
    geno(haplotype-1 genome)--&gt;|simulate snpindel -n 100000 -z 0.5|hap(snpindel.vcf)
    hap--&gt;hap1(snpindel.hap1.vcf)
    hap--&gt;hap2(snpindel.hap2.vcf)</pre></div>
<doc-anchor-target id="step-4">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#step-4">#</doc-anchor-trigger>
        <span>Step 4</span>
    </h4>
</doc-anchor-target>
<p>Use the resulting haplotype VCFs to simulate known variants onto the <strong>haplotype genomes</strong> from
<doc-anchor-trigger to="#step-2">Step 2</doc-anchor-trigger>.</p>
<div class="overflow-hidden mb-6 py-4 dark:px-5 dark:bg-white rounded-md"><pre class="mermaid w-full">graph LR
    subgraph id1 [&quot;Haplotype 1 inputs&quot;]
    hap1(snpindel.hap1.vcf)---geno(haplotype-1 genome)
    end
    id1--&gt;|simulate inversion -v|genohap1(haplotype-1 genome with new variants)
    style id1 fill:#f0f0f0,stroke:#e8e8e8,stroke-width:2px</pre></div>
<div class="overflow-hidden mb-6 py-4 dark:px-5 dark:bg-white rounded-md"><pre class="mermaid w-full">graph LR
    subgraph id2 [&quot;Haplotype 2 inputs&quot;]
    hap1(snpindel.hap2.vcf)---geno(haplotype-2 genome)
    end
    id2--&gt;|simulate inversion -v|genohap2(haplotype-2 genome with new variants)
    style id2 fill:#f0f0f0,stroke:#e8e8e8,stroke-width:2px</pre></div>
<doc-anchor-target id="step-5">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#step-5">#</doc-anchor-trigger>
        <span>Step 5</span>
    </h4>
</doc-anchor-target>
<p>Repeat <doc-anchor-trigger to="#step-3"><strong>Step 3</strong></doc-anchor-trigger> and <doc-anchor-trigger to="#step-4"><strong>Step 4</strong></doc-anchor-trigger> to your heart&#x27;s content.</p>

                                
                                <!-- Required only on API pages -->
                                <doc-toolbar-member-filter-no-results></doc-toolbar-member-filter-no-results>
                            </div>
                            <footer class="clear-both">
                                <div class="flex flex-wrap items-center justify-between mt-14">
                                    <a class="my-2.5 inline-flex items-center text-sm whitespace-nowrap text-blue-500 dark:text-blue-400 hover:text-blue-700 hover:underline" href="https://github.com/pdimens/HARPY/edit/docs/Modules/Simulate/simulate-variants.md" target="_blank" rel="noopener">
                                        <svg class="mr-1.5" xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M20 12c-.55 0-1 .45-1 1v7c0 .55-.45 1-1 1H4c-.55 0-1-.45-1-1V6c0-.55.45-1 1-1h7c.55 0 1-.45 1-1s-.45-1-1-1H4C2.35 3 1 4.35 1 6v14c0 1.65 1.35 3 3 3h14c1.65 0 3-1.35 3-3v-7c0-.55-.45-1-1-1z" /><path d="M22.21 1.79c-1.18-1.18-3.24-1.18-4.41 0l-9.5 9.5c-.13.13-.22.29-.26.46l-1 4c-.08.34.01.7.26.95.18.2.44.3.7.3.08 0 .16-.01.24-.03l4-1c.18-.04.34-.13.46-.26l9.5-9.5c1.22-1.22 1.22-3.2.01-4.42zm-1.42 3l-9.3 9.3-2.11.53.53-2.11 9.3-9.3c.42-.42 1.16-.42 1.59 0 .43.43.43 1.15-.01 1.58z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        <span>Edit this page</span>
                                    </a>
                                </div>
                            
                                <nav class="flex mt-14">
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 h-full flex items-center break-normal font-medium text-blue-500 dark:text-blue-400 border border-gray-300 hover:border-gray-400 dark:border-dark-650 dark:hover:border-dark-450 rounded-l-lg transition-colors duration-150 relative hover:z-5" href="../../../modules/align/minimap/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mr-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19 11H7.41l5.29-5.29a.996.996 0 10-1.41-1.41l-7 7a1 1 0 000 1.42l7 7a1.024 1.024 0 001.42-.01.996.996 0 000-1.41L7.41 13H19c.55 0 1-.45 1-1s-.45-1-1-1z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                            <span>
                                                <span class="block text-xs font-normal text-gray-400 dark:text-dark-400">Previous</span>
                                                <span class="block mt-1">Minimap</span>
                                            </span>
                                        </a>
                                    </div>
                            
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 -mx-px h-full flex items-center justify-end break-normal font-medium text-blue-500 dark:text-blue-400 border border-gray-300 hover:border-gray-400 dark:border-dark-650 dark:hover:border-dark-450 rounded-r-lg transition-colors duration-150 relative hover:z-5" href="../../../modules/snp/">
                                            <span>
                                                <span class="block text-xs font-normal text-right text-gray-400 dark:text-dark-400">Next</span>
                                                <span class="block mt-1">SNP</span>
                                            </span>
                                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19.92 12.38a1 1 0 00-.22-1.09l-7-7a.996.996 0 10-1.41 1.41l5.3 5.3H5c-.55 0-1 .45-1 1s.45 1 1 1h11.59l-5.29 5.29a.996.996 0 000 1.41c.19.2.44.3.7.3s.51-.1.71-.29l7-7c.09-.09.16-.21.21-.33z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        </a>
                                    </div>
                                </nav>
                            </footer>
                        </main>
                
                        <div class="border-t dark:border-dark-650 pt-6 mb-8">
                            <footer class="flex flex-wrap items-center justify-between">
                                <div>
                                    <ul class="flex flex-wrap items-center text-sm">
                                    </ul>
                                </div>
                                <div class="docs-copyright py-2 text-gray-500 dark:text-dark-350 text-sm leading-relaxed"><p> Copyright 2024. All rights reserved.</p></div>
                            </footer>
                        </div>
                    </div>
                
                    <!-- Rendered if sidebar right is enabled -->
                    <!-- Sidebar right skeleton-->
                    <div v-cloak class="fixed top-0 bottom-0 right-0 translate-x-full bg-white border-gray-200 lg:sticky lg:border-l lg:shrink-0 lg:pt-6 lg:transform-none sm:w-1/2 lg:w-64 lg:z-0 md:w-104 sidebar-right skeleton dark:bg-dark-850 dark:border-dark-650">
                        <div class="pl-5">
                            <div class="w-32 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                            <div class="w-48 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                            <div class="w-40 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                        </div>
                    </div>
                
                    <!-- User should be able to hide sidebar right -->
                    <doc-sidebar-right v-cloak></doc-sidebar-right>
                </div>

            </div>
        </div>
    
        <doc-search-mobile></doc-search-mobile>
        <doc-back-to-top></doc-back-to-top>
    </div>


    <div id="docs-overlay-target"></div>

    <script data-cfasync="false">window.__DOCS__ = { "title": "Variants", level: 3, icon: "file", hasPrism: true, hasMermaid: true, hasMath: false, tocDepth: 23 }</script>
</body>
</html>
